{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "8fd119b3",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "----------------------------debug base_layer.py 983\n",
      "str(self.name): block1_conv1\n",
      "The caller function is 'error_handler' in module 'keras.utils.traceback_utils' (current in base_layer.py 989)\n",
      "args: (<KerasTensor: shape=(None, 224, 224, 3) dtype=float32 (created by layer 'input_1')>,)\n",
      "kwargs: {}\n",
      "----------------------------debug base_layer.py 1009\n",
      "----------------------------debug base_layer.py 1011\n",
      "----------------------------debug base_layer.py 1013\n",
      "----------------------------debug base_layer.py 1023\n",
      "The caller function is '_functional_construction_call' in module 'keras.engine.base_layer' (current in base_layer.py 2358)\n",
      "The caller function is '<listcomp>' in module 'tensorflow.python.util.nest' (currently in keras_tensor.py 670)\n",
      "----------------------------debug keras_tensor.py 243\n",
      "The caller function is 'keras_tensor_to_placeholder' in module 'keras.engine.keras_tensor' (currently in keras_tensor.py 248)\n",
      "keras_tensor.py self: KerasTensor(type_spec=TensorSpec(shape=(None, 224, 224, 3), dtype=tf.float32, name='input_1'), name='input_1', description=\"created by layer 'input_1'\")\n",
      "component_to_placeholder: <function KerasTensor._to_placeholder.<locals>.component_to_placeholder at 0x0000013B7E2BB7F0>\n",
      "self.type_spec: TensorSpec(shape=(None, 224, 224, 3), dtype=tf.float32, name='input_1')\n",
      "The caller function is '<listcomp>' in module 'tensorflow.python.util.nest' (current in keras.engine.keras_tensor 238)\n",
      "The caller function is 'placeholder' in module 'tensorflow.python.ops.array_ops' (current in gen_array_ops.py 6902)\n",
      "name: None\n",
      "dtype: 1\n",
      "The caller function is '<listcomp>' in module 'tensorflow.python.util.nest' (currently in keras_tensor.py 670)\n",
      "----------------------------debug base_layer.py 2441\n",
      "----------------------------debug traceback_utils.py 101\n",
      "The caller function is '_infer_output_signature' in module 'keras.engine.base_layer' (current in base_layer.py 989)\n",
      "----------------------------debug base_layer.py 2462\n",
      "----------------------------debug base_layer.py 2471\n",
      "----------------------------debug base_layer.py 983\n",
      "str(self.name): block1_conv2\n",
      "The caller function is 'error_handler' in module 'keras.utils.traceback_utils' (current in base_layer.py 989)\n",
      "args: (<KerasTensor: shape=(None, 224, 224, 64) dtype=float32 (created by layer 'block1_conv1')>,)\n",
      "kwargs: {}\n",
      "----------------------------debug base_layer.py 1009\n",
      "----------------------------debug base_layer.py 1011\n",
      "----------------------------debug base_layer.py 1013\n",
      "----------------------------debug base_layer.py 1023\n",
      "The caller function is '_functional_construction_call' in module 'keras.engine.base_layer' (current in base_layer.py 2358)\n",
      "The caller function is '<listcomp>' in module 'tensorflow.python.util.nest' (currently in keras_tensor.py 670)\n",
      "----------------------------debug keras_tensor.py 243\n",
      "The caller function is 'keras_tensor_to_placeholder' in module 'keras.engine.keras_tensor' (currently in keras_tensor.py 248)\n",
      "keras_tensor.py self: KerasTensor(type_spec=TensorSpec(shape=(None, 224, 224, 64), dtype=tf.float32, name=None), name='block1_conv1/Relu:0', description=\"created by layer 'block1_conv1'\")\n",
      "component_to_placeholder: <function KerasTensor._to_placeholder.<locals>.component_to_placeholder at 0x0000013B7E2BB6D0>\n",
      "self.type_spec: TensorSpec(shape=(None, 224, 224, 64), dtype=tf.float32, name=None)\n",
      "The caller function is '<listcomp>' in module 'tensorflow.python.util.nest' (current in keras.engine.keras_tensor 238)\n",
      "The caller function is 'placeholder' in module 'tensorflow.python.ops.array_ops' (current in gen_array_ops.py 6902)\n",
      "name: None\n",
      "dtype: 1\n",
      "The caller function is '<listcomp>' in module 'tensorflow.python.util.nest' (currently in keras_tensor.py 670)\n",
      "----------------------------debug base_layer.py 2441\n",
      "----------------------------debug traceback_utils.py 101\n",
      "The caller function is '_infer_output_signature' in module 'keras.engine.base_layer' (current in base_layer.py 989)\n",
      "----------------------------debug base_layer.py 2462\n",
      "----------------------------debug base_layer.py 2471\n",
      "----------------------------debug base_layer.py 983\n",
      "str(self.name): block1_pool\n",
      "The caller function is 'error_handler' in module 'keras.utils.traceback_utils' (current in base_layer.py 989)\n",
      "args: (<KerasTensor: shape=(None, 224, 224, 64) dtype=float32 (created by layer 'block1_conv2')>,)\n",
      "kwargs: {}\n",
      "----------------------------debug base_layer.py 1009\n",
      "----------------------------debug base_layer.py 1011\n",
      "----------------------------debug base_layer.py 1013\n",
      "----------------------------debug base_layer.py 1023\n",
      "The caller function is '_functional_construction_call' in module 'keras.engine.base_layer' (current in base_layer.py 2358)\n",
      "The caller function is '<listcomp>' in module 'tensorflow.python.util.nest' (currently in keras_tensor.py 670)\n",
      "----------------------------debug keras_tensor.py 243\n",
      "The caller function is 'keras_tensor_to_placeholder' in module 'keras.engine.keras_tensor' (currently in keras_tensor.py 248)\n",
      "keras_tensor.py self: KerasTensor(type_spec=TensorSpec(shape=(None, 224, 224, 64), dtype=tf.float32, name=None), name='block1_conv2/Relu:0', description=\"created by layer 'block1_conv2'\")\n",
      "component_to_placeholder: <function KerasTensor._to_placeholder.<locals>.component_to_placeholder at 0x0000013B7E2BB760>\n",
      "self.type_spec: TensorSpec(shape=(None, 224, 224, 64), dtype=tf.float32, name=None)\n",
      "The caller function is '<listcomp>' in module 'tensorflow.python.util.nest' (current in keras.engine.keras_tensor 238)\n",
      "The caller function is 'placeholder' in module 'tensorflow.python.ops.array_ops' (current in gen_array_ops.py 6902)\n",
      "name: None\n",
      "dtype: 1\n",
      "The caller function is '<listcomp>' in module 'tensorflow.python.util.nest' (currently in keras_tensor.py 670)\n",
      "----------------------------debug base_layer.py 2441\n",
      "----------------------------debug traceback_utils.py 101\n",
      "The caller function is '_infer_output_signature' in module 'keras.engine.base_layer' (current in base_layer.py 989)\n",
      "----------------------------debug base_layer.py 2462\n",
      "----------------------------debug base_layer.py 2471\n",
      "----------------------------debug base_layer.py 983\n",
      "str(self.name): block2_conv1\n",
      "The caller function is 'error_handler' in module 'keras.utils.traceback_utils' (current in base_layer.py 989)\n",
      "args: (<KerasTensor: shape=(None, 112, 112, 64) dtype=float32 (created by layer 'block1_pool')>,)\n",
      "kwargs: {}\n",
      "----------------------------debug base_layer.py 1009\n",
      "----------------------------debug base_layer.py 1011\n",
      "----------------------------debug base_layer.py 1013\n",
      "----------------------------debug base_layer.py 1023\n",
      "The caller function is '_functional_construction_call' in module 'keras.engine.base_layer' (current in base_layer.py 2358)\n",
      "The caller function is '<listcomp>' in module 'tensorflow.python.util.nest' (currently in keras_tensor.py 670)\n",
      "----------------------------debug keras_tensor.py 243\n",
      "The caller function is 'keras_tensor_to_placeholder' in module 'keras.engine.keras_tensor' (currently in keras_tensor.py 248)\n",
      "keras_tensor.py self: KerasTensor(type_spec=TensorSpec(shape=(None, 112, 112, 64), dtype=tf.float32, name=None), name='block1_pool/MaxPool:0', description=\"created by layer 'block1_pool'\")\n",
      "component_to_placeholder: <function KerasTensor._to_placeholder.<locals>.component_to_placeholder at 0x0000013B7E2BB910>\n",
      "self.type_spec: TensorSpec(shape=(None, 112, 112, 64), dtype=tf.float32, name=None)\n",
      "The caller function is '<listcomp>' in module 'tensorflow.python.util.nest' (current in keras.engine.keras_tensor 238)\n",
      "The caller function is 'placeholder' in module 'tensorflow.python.ops.array_ops' (current in gen_array_ops.py 6902)\n",
      "name: None\n",
      "dtype: 1\n",
      "The caller function is '<listcomp>' in module 'tensorflow.python.util.nest' (currently in keras_tensor.py 670)\n",
      "----------------------------debug base_layer.py 2441\n",
      "----------------------------debug traceback_utils.py 101\n",
      "The caller function is '_infer_output_signature' in module 'keras.engine.base_layer' (current in base_layer.py 989)\n",
      "----------------------------debug base_layer.py 2462\n",
      "----------------------------debug base_layer.py 2471\n",
      "----------------------------debug base_layer.py 983\n",
      "str(self.name): block2_conv2\n",
      "The caller function is 'error_handler' in module 'keras.utils.traceback_utils' (current in base_layer.py 989)\n",
      "args: (<KerasTensor: shape=(None, 112, 112, 128) dtype=float32 (created by layer 'block2_conv1')>,)\n",
      "kwargs: {}\n",
      "----------------------------debug base_layer.py 1009\n",
      "----------------------------debug base_layer.py 1011\n",
      "----------------------------debug base_layer.py 1013\n",
      "----------------------------debug base_layer.py 1023\n",
      "The caller function is '_functional_construction_call' in module 'keras.engine.base_layer' (current in base_layer.py 2358)\n",
      "The caller function is '<listcomp>' in module 'tensorflow.python.util.nest' (currently in keras_tensor.py 670)\n",
      "----------------------------debug keras_tensor.py 243\n",
      "The caller function is 'keras_tensor_to_placeholder' in module 'keras.engine.keras_tensor' (currently in keras_tensor.py 248)\n",
      "keras_tensor.py self: KerasTensor(type_spec=TensorSpec(shape=(None, 112, 112, 128), dtype=tf.float32, name=None), name='block2_conv1/Relu:0', description=\"created by layer 'block2_conv1'\")\n",
      "component_to_placeholder: <function KerasTensor._to_placeholder.<locals>.component_to_placeholder at 0x0000013B7E2BB6D0>\n",
      "self.type_spec: TensorSpec(shape=(None, 112, 112, 128), dtype=tf.float32, name=None)\n",
      "The caller function is '<listcomp>' in module 'tensorflow.python.util.nest' (current in keras.engine.keras_tensor 238)\n",
      "The caller function is 'placeholder' in module 'tensorflow.python.ops.array_ops' (current in gen_array_ops.py 6902)\n",
      "name: None\n",
      "dtype: 1\n",
      "The caller function is '<listcomp>' in module 'tensorflow.python.util.nest' (currently in keras_tensor.py 670)\n",
      "----------------------------debug base_layer.py 2441\n",
      "----------------------------debug traceback_utils.py 101\n",
      "The caller function is '_infer_output_signature' in module 'keras.engine.base_layer' (current in base_layer.py 989)\n",
      "----------------------------debug base_layer.py 2462\n",
      "----------------------------debug base_layer.py 2471\n",
      "----------------------------debug base_layer.py 983\n",
      "str(self.name): block2_pool\n",
      "The caller function is 'error_handler' in module 'keras.utils.traceback_utils' (current in base_layer.py 989)\n",
      "args: (<KerasTensor: shape=(None, 112, 112, 128) dtype=float32 (created by layer 'block2_conv2')>,)\n",
      "kwargs: {}\n",
      "----------------------------debug base_layer.py 1009\n",
      "----------------------------debug base_layer.py 1011\n",
      "----------------------------debug base_layer.py 1013\n",
      "----------------------------debug base_layer.py 1023\n",
      "The caller function is '_functional_construction_call' in module 'keras.engine.base_layer' (current in base_layer.py 2358)\n",
      "The caller function is '<listcomp>' in module 'tensorflow.python.util.nest' (currently in keras_tensor.py 670)\n",
      "----------------------------debug keras_tensor.py 243\n",
      "The caller function is 'keras_tensor_to_placeholder' in module 'keras.engine.keras_tensor' (currently in keras_tensor.py 248)\n",
      "keras_tensor.py self: KerasTensor(type_spec=TensorSpec(shape=(None, 112, 112, 128), dtype=tf.float32, name=None), name='block2_conv2/Relu:0', description=\"created by layer 'block2_conv2'\")\n",
      "component_to_placeholder: <function KerasTensor._to_placeholder.<locals>.component_to_placeholder at 0x0000013B7E2BB760>\n",
      "self.type_spec: TensorSpec(shape=(None, 112, 112, 128), dtype=tf.float32, name=None)\n",
      "The caller function is '<listcomp>' in module 'tensorflow.python.util.nest' (current in keras.engine.keras_tensor 238)\n",
      "The caller function is 'placeholder' in module 'tensorflow.python.ops.array_ops' (current in gen_array_ops.py 6902)\n",
      "name: None\n",
      "dtype: 1\n",
      "The caller function is '<listcomp>' in module 'tensorflow.python.util.nest' (currently in keras_tensor.py 670)\n",
      "----------------------------debug base_layer.py 2441\n",
      "----------------------------debug traceback_utils.py 101\n",
      "The caller function is '_infer_output_signature' in module 'keras.engine.base_layer' (current in base_layer.py 989)\n",
      "----------------------------debug base_layer.py 2462\n",
      "----------------------------debug base_layer.py 2471\n",
      "----------------------------debug base_layer.py 983\n",
      "str(self.name): block3_conv1\n",
      "The caller function is 'error_handler' in module 'keras.utils.traceback_utils' (current in base_layer.py 989)\n",
      "args: (<KerasTensor: shape=(None, 56, 56, 128) dtype=float32 (created by layer 'block2_pool')>,)\n",
      "kwargs: {}\n",
      "----------------------------debug base_layer.py 1009\n",
      "----------------------------debug base_layer.py 1011\n",
      "----------------------------debug base_layer.py 1013\n",
      "----------------------------debug base_layer.py 1023\n",
      "The caller function is '_functional_construction_call' in module 'keras.engine.base_layer' (current in base_layer.py 2358)\n",
      "The caller function is '<listcomp>' in module 'tensorflow.python.util.nest' (currently in keras_tensor.py 670)\n",
      "----------------------------debug keras_tensor.py 243\n",
      "The caller function is 'keras_tensor_to_placeholder' in module 'keras.engine.keras_tensor' (currently in keras_tensor.py 248)\n",
      "keras_tensor.py self: KerasTensor(type_spec=TensorSpec(shape=(None, 56, 56, 128), dtype=tf.float32, name=None), name='block2_pool/MaxPool:0', description=\"created by layer 'block2_pool'\")\n",
      "component_to_placeholder: <function KerasTensor._to_placeholder.<locals>.component_to_placeholder at 0x0000013B7E2BB910>\n",
      "self.type_spec: TensorSpec(shape=(None, 56, 56, 128), dtype=tf.float32, name=None)\n",
      "The caller function is '<listcomp>' in module 'tensorflow.python.util.nest' (current in keras.engine.keras_tensor 238)\n",
      "The caller function is 'placeholder' in module 'tensorflow.python.ops.array_ops' (current in gen_array_ops.py 6902)\n",
      "name: None\n",
      "dtype: 1\n",
      "The caller function is '<listcomp>' in module 'tensorflow.python.util.nest' (currently in keras_tensor.py 670)\n",
      "----------------------------debug base_layer.py 2441\n",
      "----------------------------debug traceback_utils.py 101\n",
      "The caller function is '_infer_output_signature' in module 'keras.engine.base_layer' (current in base_layer.py 989)\n",
      "----------------------------debug base_layer.py 2462\n",
      "----------------------------debug base_layer.py 2471\n",
      "----------------------------debug base_layer.py 983\n",
      "str(self.name): block3_conv2\n",
      "The caller function is 'error_handler' in module 'keras.utils.traceback_utils' (current in base_layer.py 989)\n",
      "args: (<KerasTensor: shape=(None, 56, 56, 256) dtype=float32 (created by layer 'block3_conv1')>,)\n",
      "kwargs: {}\n",
      "----------------------------debug base_layer.py 1009\n",
      "----------------------------debug base_layer.py 1011\n",
      "----------------------------debug base_layer.py 1013\n",
      "----------------------------debug base_layer.py 1023\n",
      "The caller function is '_functional_construction_call' in module 'keras.engine.base_layer' (current in base_layer.py 2358)\n",
      "The caller function is '<listcomp>' in module 'tensorflow.python.util.nest' (currently in keras_tensor.py 670)\n",
      "----------------------------debug keras_tensor.py 243\n",
      "The caller function is 'keras_tensor_to_placeholder' in module 'keras.engine.keras_tensor' (currently in keras_tensor.py 248)\n",
      "keras_tensor.py self: KerasTensor(type_spec=TensorSpec(shape=(None, 56, 56, 256), dtype=tf.float32, name=None), name='block3_conv1/Relu:0', description=\"created by layer 'block3_conv1'\")\n",
      "component_to_placeholder: <function KerasTensor._to_placeholder.<locals>.component_to_placeholder at 0x0000013B7E2BB6D0>\n",
      "self.type_spec: TensorSpec(shape=(None, 56, 56, 256), dtype=tf.float32, name=None)\n",
      "The caller function is '<listcomp>' in module 'tensorflow.python.util.nest' (current in keras.engine.keras_tensor 238)\n",
      "The caller function is 'placeholder' in module 'tensorflow.python.ops.array_ops' (current in gen_array_ops.py 6902)\n",
      "name: None\n",
      "dtype: 1\n",
      "The caller function is '<listcomp>' in module 'tensorflow.python.util.nest' (currently in keras_tensor.py 670)\n",
      "----------------------------debug base_layer.py 2441\n",
      "----------------------------debug traceback_utils.py 101\n",
      "The caller function is '_infer_output_signature' in module 'keras.engine.base_layer' (current in base_layer.py 989)\n",
      "----------------------------debug base_layer.py 2462\n",
      "----------------------------debug base_layer.py 2471\n",
      "----------------------------debug base_layer.py 983\n",
      "str(self.name): block3_conv3\n",
      "The caller function is 'error_handler' in module 'keras.utils.traceback_utils' (current in base_layer.py 989)\n",
      "args: (<KerasTensor: shape=(None, 56, 56, 256) dtype=float32 (created by layer 'block3_conv2')>,)\n",
      "kwargs: {}\n",
      "----------------------------debug base_layer.py 1009\n",
      "----------------------------debug base_layer.py 1011\n",
      "----------------------------debug base_layer.py 1013\n",
      "----------------------------debug base_layer.py 1023\n",
      "The caller function is '_functional_construction_call' in module 'keras.engine.base_layer' (current in base_layer.py 2358)\n",
      "The caller function is '<listcomp>' in module 'tensorflow.python.util.nest' (currently in keras_tensor.py 670)\n",
      "----------------------------debug keras_tensor.py 243\n",
      "The caller function is 'keras_tensor_to_placeholder' in module 'keras.engine.keras_tensor' (currently in keras_tensor.py 248)\n",
      "keras_tensor.py self: KerasTensor(type_spec=TensorSpec(shape=(None, 56, 56, 256), dtype=tf.float32, name=None), name='block3_conv2/Relu:0', description=\"created by layer 'block3_conv2'\")\n",
      "component_to_placeholder: <function KerasTensor._to_placeholder.<locals>.component_to_placeholder at 0x0000013B7E2BB760>\n",
      "self.type_spec: TensorSpec(shape=(None, 56, 56, 256), dtype=tf.float32, name=None)\n",
      "The caller function is '<listcomp>' in module 'tensorflow.python.util.nest' (current in keras.engine.keras_tensor 238)\n",
      "The caller function is 'placeholder' in module 'tensorflow.python.ops.array_ops' (current in gen_array_ops.py 6902)\n",
      "name: None\n",
      "dtype: 1\n",
      "The caller function is '<listcomp>' in module 'tensorflow.python.util.nest' (currently in keras_tensor.py 670)\n",
      "----------------------------debug base_layer.py 2441\n",
      "----------------------------debug traceback_utils.py 101\n",
      "The caller function is '_infer_output_signature' in module 'keras.engine.base_layer' (current in base_layer.py 989)\n",
      "----------------------------debug base_layer.py 2462\n",
      "----------------------------debug base_layer.py 2471\n",
      "----------------------------debug base_layer.py 983\n",
      "str(self.name): block3_conv4\n",
      "The caller function is 'error_handler' in module 'keras.utils.traceback_utils' (current in base_layer.py 989)\n",
      "args: (<KerasTensor: shape=(None, 56, 56, 256) dtype=float32 (created by layer 'block3_conv3')>,)\n",
      "kwargs: {}\n",
      "----------------------------debug base_layer.py 1009\n",
      "----------------------------debug base_layer.py 1011\n",
      "----------------------------debug base_layer.py 1013\n",
      "----------------------------debug base_layer.py 1023\n",
      "The caller function is '_functional_construction_call' in module 'keras.engine.base_layer' (current in base_layer.py 2358)\n",
      "The caller function is '<listcomp>' in module 'tensorflow.python.util.nest' (currently in keras_tensor.py 670)\n",
      "----------------------------debug keras_tensor.py 243\n",
      "The caller function is 'keras_tensor_to_placeholder' in module 'keras.engine.keras_tensor' (currently in keras_tensor.py 248)\n",
      "keras_tensor.py self: KerasTensor(type_spec=TensorSpec(shape=(None, 56, 56, 256), dtype=tf.float32, name=None), name='block3_conv3/Relu:0', description=\"created by layer 'block3_conv3'\")\n",
      "component_to_placeholder: <function KerasTensor._to_placeholder.<locals>.component_to_placeholder at 0x0000013B7E2BBD00>\n",
      "self.type_spec: TensorSpec(shape=(None, 56, 56, 256), dtype=tf.float32, name=None)\n",
      "The caller function is '<listcomp>' in module 'tensorflow.python.util.nest' (current in keras.engine.keras_tensor 238)\n",
      "The caller function is 'placeholder' in module 'tensorflow.python.ops.array_ops' (current in gen_array_ops.py 6902)\n",
      "name: None\n",
      "dtype: 1\n",
      "The caller function is '<listcomp>' in module 'tensorflow.python.util.nest' (currently in keras_tensor.py 670)\n",
      "----------------------------debug base_layer.py 2441\n",
      "----------------------------debug traceback_utils.py 101\n",
      "The caller function is '_infer_output_signature' in module 'keras.engine.base_layer' (current in base_layer.py 989)\n",
      "----------------------------debug base_layer.py 2462\n",
      "----------------------------debug base_layer.py 2471\n",
      "----------------------------debug base_layer.py 983\n",
      "str(self.name): block3_pool\n",
      "The caller function is 'error_handler' in module 'keras.utils.traceback_utils' (current in base_layer.py 989)\n",
      "args: (<KerasTensor: shape=(None, 56, 56, 256) dtype=float32 (created by layer 'block3_conv4')>,)\n",
      "kwargs: {}\n",
      "----------------------------debug base_layer.py 1009\n",
      "----------------------------debug base_layer.py 1011\n",
      "----------------------------debug base_layer.py 1013\n",
      "----------------------------debug base_layer.py 1023\n",
      "The caller function is '_functional_construction_call' in module 'keras.engine.base_layer' (current in base_layer.py 2358)\n",
      "The caller function is '<listcomp>' in module 'tensorflow.python.util.nest' (currently in keras_tensor.py 670)\n",
      "----------------------------debug keras_tensor.py 243\n",
      "The caller function is 'keras_tensor_to_placeholder' in module 'keras.engine.keras_tensor' (currently in keras_tensor.py 248)\n",
      "keras_tensor.py self: KerasTensor(type_spec=TensorSpec(shape=(None, 56, 56, 256), dtype=tf.float32, name=None), name='block3_conv4/Relu:0', description=\"created by layer 'block3_conv4'\")\n",
      "component_to_placeholder: <function KerasTensor._to_placeholder.<locals>.component_to_placeholder at 0x0000013B7E2BB6D0>\n",
      "self.type_spec: TensorSpec(shape=(None, 56, 56, 256), dtype=tf.float32, name=None)\n",
      "The caller function is '<listcomp>' in module 'tensorflow.python.util.nest' (current in keras.engine.keras_tensor 238)\n",
      "The caller function is 'placeholder' in module 'tensorflow.python.ops.array_ops' (current in gen_array_ops.py 6902)\n",
      "name: None\n",
      "dtype: 1\n",
      "The caller function is '<listcomp>' in module 'tensorflow.python.util.nest' (currently in keras_tensor.py 670)\n",
      "----------------------------debug base_layer.py 2441\n",
      "----------------------------debug traceback_utils.py 101\n",
      "The caller function is '_infer_output_signature' in module 'keras.engine.base_layer' (current in base_layer.py 989)\n",
      "----------------------------debug base_layer.py 2462\n",
      "----------------------------debug base_layer.py 2471\n",
      "----------------------------debug base_layer.py 983\n",
      "str(self.name): block4_conv1\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The caller function is 'error_handler' in module 'keras.utils.traceback_utils' (current in base_layer.py 989)\n",
      "args: (<KerasTensor: shape=(None, 28, 28, 256) dtype=float32 (created by layer 'block3_pool')>,)\n",
      "kwargs: {}\n",
      "----------------------------debug base_layer.py 1009\n",
      "----------------------------debug base_layer.py 1011\n",
      "----------------------------debug base_layer.py 1013\n",
      "----------------------------debug base_layer.py 1023\n",
      "The caller function is '_functional_construction_call' in module 'keras.engine.base_layer' (current in base_layer.py 2358)\n",
      "The caller function is '<listcomp>' in module 'tensorflow.python.util.nest' (currently in keras_tensor.py 670)\n",
      "----------------------------debug keras_tensor.py 243\n",
      "The caller function is 'keras_tensor_to_placeholder' in module 'keras.engine.keras_tensor' (currently in keras_tensor.py 248)\n",
      "keras_tensor.py self: KerasTensor(type_spec=TensorSpec(shape=(None, 28, 28, 256), dtype=tf.float32, name=None), name='block3_pool/MaxPool:0', description=\"created by layer 'block3_pool'\")\n",
      "component_to_placeholder: <function KerasTensor._to_placeholder.<locals>.component_to_placeholder at 0x0000013B7E2BB9A0>\n",
      "self.type_spec: TensorSpec(shape=(None, 28, 28, 256), dtype=tf.float32, name=None)\n",
      "The caller function is '<listcomp>' in module 'tensorflow.python.util.nest' (current in keras.engine.keras_tensor 238)\n",
      "The caller function is 'placeholder' in module 'tensorflow.python.ops.array_ops' (current in gen_array_ops.py 6902)\n",
      "name: None\n",
      "dtype: 1\n",
      "The caller function is '<listcomp>' in module 'tensorflow.python.util.nest' (currently in keras_tensor.py 670)\n",
      "----------------------------debug base_layer.py 2441\n",
      "----------------------------debug traceback_utils.py 101\n",
      "The caller function is '_infer_output_signature' in module 'keras.engine.base_layer' (current in base_layer.py 989)\n",
      "----------------------------debug base_layer.py 2462\n",
      "----------------------------debug base_layer.py 2471\n",
      "----------------------------debug base_layer.py 983\n",
      "str(self.name): block4_conv2\n",
      "The caller function is 'error_handler' in module 'keras.utils.traceback_utils' (current in base_layer.py 989)\n",
      "args: (<KerasTensor: shape=(None, 28, 28, 512) dtype=float32 (created by layer 'block4_conv1')>,)\n",
      "kwargs: {}\n",
      "----------------------------debug base_layer.py 1009\n",
      "----------------------------debug base_layer.py 1011\n",
      "----------------------------debug base_layer.py 1013\n",
      "----------------------------debug base_layer.py 1023\n",
      "The caller function is '_functional_construction_call' in module 'keras.engine.base_layer' (current in base_layer.py 2358)\n",
      "The caller function is '<listcomp>' in module 'tensorflow.python.util.nest' (currently in keras_tensor.py 670)\n",
      "----------------------------debug keras_tensor.py 243\n",
      "The caller function is 'keras_tensor_to_placeholder' in module 'keras.engine.keras_tensor' (currently in keras_tensor.py 248)\n",
      "keras_tensor.py self: KerasTensor(type_spec=TensorSpec(shape=(None, 28, 28, 512), dtype=tf.float32, name=None), name='block4_conv1/Relu:0', description=\"created by layer 'block4_conv1'\")\n",
      "component_to_placeholder: <function KerasTensor._to_placeholder.<locals>.component_to_placeholder at 0x0000013B7E2BBA30>\n",
      "self.type_spec: TensorSpec(shape=(None, 28, 28, 512), dtype=tf.float32, name=None)\n",
      "The caller function is '<listcomp>' in module 'tensorflow.python.util.nest' (current in keras.engine.keras_tensor 238)\n",
      "The caller function is 'placeholder' in module 'tensorflow.python.ops.array_ops' (current in gen_array_ops.py 6902)\n",
      "name: None\n",
      "dtype: 1\n",
      "The caller function is '<listcomp>' in module 'tensorflow.python.util.nest' (currently in keras_tensor.py 670)\n",
      "----------------------------debug base_layer.py 2441\n",
      "----------------------------debug traceback_utils.py 101\n",
      "The caller function is '_infer_output_signature' in module 'keras.engine.base_layer' (current in base_layer.py 989)\n",
      "----------------------------debug base_layer.py 2462\n",
      "----------------------------debug base_layer.py 2471\n",
      "----------------------------debug base_layer.py 983\n",
      "str(self.name): block4_conv3\n",
      "The caller function is 'error_handler' in module 'keras.utils.traceback_utils' (current in base_layer.py 989)\n",
      "args: (<KerasTensor: shape=(None, 28, 28, 512) dtype=float32 (created by layer 'block4_conv2')>,)\n",
      "kwargs: {}\n",
      "----------------------------debug base_layer.py 1009\n",
      "----------------------------debug base_layer.py 1011\n",
      "----------------------------debug base_layer.py 1013\n",
      "----------------------------debug base_layer.py 1023\n",
      "The caller function is '_functional_construction_call' in module 'keras.engine.base_layer' (current in base_layer.py 2358)\n",
      "The caller function is '<listcomp>' in module 'tensorflow.python.util.nest' (currently in keras_tensor.py 670)\n",
      "----------------------------debug keras_tensor.py 243\n",
      "The caller function is 'keras_tensor_to_placeholder' in module 'keras.engine.keras_tensor' (currently in keras_tensor.py 248)\n",
      "keras_tensor.py self: KerasTensor(type_spec=TensorSpec(shape=(None, 28, 28, 512), dtype=tf.float32, name=None), name='block4_conv2/Relu:0', description=\"created by layer 'block4_conv2'\")\n",
      "component_to_placeholder: <function KerasTensor._to_placeholder.<locals>.component_to_placeholder at 0x0000013B7E2BB9A0>\n",
      "self.type_spec: TensorSpec(shape=(None, 28, 28, 512), dtype=tf.float32, name=None)\n",
      "The caller function is '<listcomp>' in module 'tensorflow.python.util.nest' (current in keras.engine.keras_tensor 238)\n",
      "The caller function is 'placeholder' in module 'tensorflow.python.ops.array_ops' (current in gen_array_ops.py 6902)\n",
      "name: None\n",
      "dtype: 1\n",
      "The caller function is '<listcomp>' in module 'tensorflow.python.util.nest' (currently in keras_tensor.py 670)\n",
      "----------------------------debug base_layer.py 2441\n",
      "----------------------------debug traceback_utils.py 101\n",
      "The caller function is '_infer_output_signature' in module 'keras.engine.base_layer' (current in base_layer.py 989)\n",
      "----------------------------debug base_layer.py 2462\n",
      "----------------------------debug base_layer.py 2471\n",
      "----------------------------debug base_layer.py 983\n",
      "str(self.name): block4_conv4\n",
      "The caller function is 'error_handler' in module 'keras.utils.traceback_utils' (current in base_layer.py 989)\n",
      "args: (<KerasTensor: shape=(None, 28, 28, 512) dtype=float32 (created by layer 'block4_conv3')>,)\n",
      "kwargs: {}\n",
      "----------------------------debug base_layer.py 1009\n",
      "----------------------------debug base_layer.py 1011\n",
      "----------------------------debug base_layer.py 1013\n",
      "----------------------------debug base_layer.py 1023\n",
      "The caller function is '_functional_construction_call' in module 'keras.engine.base_layer' (current in base_layer.py 2358)\n",
      "The caller function is '<listcomp>' in module 'tensorflow.python.util.nest' (currently in keras_tensor.py 670)\n",
      "----------------------------debug keras_tensor.py 243\n",
      "The caller function is 'keras_tensor_to_placeholder' in module 'keras.engine.keras_tensor' (currently in keras_tensor.py 248)\n",
      "keras_tensor.py self: KerasTensor(type_spec=TensorSpec(shape=(None, 28, 28, 512), dtype=tf.float32, name=None), name='block4_conv3/Relu:0', description=\"created by layer 'block4_conv3'\")\n",
      "component_to_placeholder: <function KerasTensor._to_placeholder.<locals>.component_to_placeholder at 0x0000013B7E2BBA30>\n",
      "self.type_spec: TensorSpec(shape=(None, 28, 28, 512), dtype=tf.float32, name=None)\n",
      "The caller function is '<listcomp>' in module 'tensorflow.python.util.nest' (current in keras.engine.keras_tensor 238)\n",
      "The caller function is 'placeholder' in module 'tensorflow.python.ops.array_ops' (current in gen_array_ops.py 6902)\n",
      "name: None\n",
      "dtype: 1\n",
      "The caller function is '<listcomp>' in module 'tensorflow.python.util.nest' (currently in keras_tensor.py 670)\n",
      "----------------------------debug base_layer.py 2441\n",
      "----------------------------debug traceback_utils.py 101\n",
      "The caller function is '_infer_output_signature' in module 'keras.engine.base_layer' (current in base_layer.py 989)\n",
      "----------------------------debug base_layer.py 2462\n",
      "----------------------------debug base_layer.py 2471\n",
      "----------------------------debug base_layer.py 983\n",
      "str(self.name): block4_pool\n",
      "The caller function is 'error_handler' in module 'keras.utils.traceback_utils' (current in base_layer.py 989)\n",
      "args: (<KerasTensor: shape=(None, 28, 28, 512) dtype=float32 (created by layer 'block4_conv4')>,)\n",
      "kwargs: {}\n",
      "----------------------------debug base_layer.py 1009\n",
      "----------------------------debug base_layer.py 1011\n",
      "----------------------------debug base_layer.py 1013\n",
      "----------------------------debug base_layer.py 1023\n",
      "The caller function is '_functional_construction_call' in module 'keras.engine.base_layer' (current in base_layer.py 2358)\n",
      "The caller function is '<listcomp>' in module 'tensorflow.python.util.nest' (currently in keras_tensor.py 670)\n",
      "----------------------------debug keras_tensor.py 243\n",
      "The caller function is 'keras_tensor_to_placeholder' in module 'keras.engine.keras_tensor' (currently in keras_tensor.py 248)\n",
      "keras_tensor.py self: KerasTensor(type_spec=TensorSpec(shape=(None, 28, 28, 512), dtype=tf.float32, name=None), name='block4_conv4/Relu:0', description=\"created by layer 'block4_conv4'\")\n",
      "component_to_placeholder: <function KerasTensor._to_placeholder.<locals>.component_to_placeholder at 0x0000013B7E2BB9A0>\n",
      "self.type_spec: TensorSpec(shape=(None, 28, 28, 512), dtype=tf.float32, name=None)\n",
      "The caller function is '<listcomp>' in module 'tensorflow.python.util.nest' (current in keras.engine.keras_tensor 238)\n",
      "The caller function is 'placeholder' in module 'tensorflow.python.ops.array_ops' (current in gen_array_ops.py 6902)\n",
      "name: None\n",
      "dtype: 1\n",
      "The caller function is '<listcomp>' in module 'tensorflow.python.util.nest' (currently in keras_tensor.py 670)\n",
      "----------------------------debug base_layer.py 2441\n",
      "----------------------------debug traceback_utils.py 101\n",
      "The caller function is '_infer_output_signature' in module 'keras.engine.base_layer' (current in base_layer.py 989)\n",
      "----------------------------debug base_layer.py 2462\n",
      "----------------------------debug base_layer.py 2471\n",
      "----------------------------debug base_layer.py 983\n",
      "str(self.name): block5_conv1\n",
      "The caller function is 'error_handler' in module 'keras.utils.traceback_utils' (current in base_layer.py 989)\n",
      "args: (<KerasTensor: shape=(None, 14, 14, 512) dtype=float32 (created by layer 'block4_pool')>,)\n",
      "kwargs: {}\n",
      "----------------------------debug base_layer.py 1009\n",
      "----------------------------debug base_layer.py 1011\n",
      "----------------------------debug base_layer.py 1013\n",
      "----------------------------debug base_layer.py 1023\n",
      "The caller function is '_functional_construction_call' in module 'keras.engine.base_layer' (current in base_layer.py 2358)\n",
      "The caller function is '<listcomp>' in module 'tensorflow.python.util.nest' (currently in keras_tensor.py 670)\n",
      "----------------------------debug keras_tensor.py 243\n",
      "The caller function is 'keras_tensor_to_placeholder' in module 'keras.engine.keras_tensor' (currently in keras_tensor.py 248)\n",
      "keras_tensor.py self: KerasTensor(type_spec=TensorSpec(shape=(None, 14, 14, 512), dtype=tf.float32, name=None), name='block4_pool/MaxPool:0', description=\"created by layer 'block4_pool'\")\n",
      "component_to_placeholder: <function KerasTensor._to_placeholder.<locals>.component_to_placeholder at 0x0000013B7E2BBA30>\n",
      "self.type_spec: TensorSpec(shape=(None, 14, 14, 512), dtype=tf.float32, name=None)\n",
      "The caller function is '<listcomp>' in module 'tensorflow.python.util.nest' (current in keras.engine.keras_tensor 238)\n",
      "The caller function is 'placeholder' in module 'tensorflow.python.ops.array_ops' (current in gen_array_ops.py 6902)\n",
      "name: None\n",
      "dtype: 1\n",
      "The caller function is '<listcomp>' in module 'tensorflow.python.util.nest' (currently in keras_tensor.py 670)\n",
      "----------------------------debug base_layer.py 2441\n",
      "----------------------------debug traceback_utils.py 101\n",
      "The caller function is '_infer_output_signature' in module 'keras.engine.base_layer' (current in base_layer.py 989)\n",
      "----------------------------debug base_layer.py 2462\n",
      "----------------------------debug base_layer.py 2471\n",
      "----------------------------debug base_layer.py 983\n",
      "str(self.name): block5_conv2\n",
      "The caller function is 'error_handler' in module 'keras.utils.traceback_utils' (current in base_layer.py 989)\n",
      "args: (<KerasTensor: shape=(None, 14, 14, 512) dtype=float32 (created by layer 'block5_conv1')>,)\n",
      "kwargs: {}\n",
      "----------------------------debug base_layer.py 1009\n",
      "----------------------------debug base_layer.py 1011\n",
      "----------------------------debug base_layer.py 1013\n",
      "----------------------------debug base_layer.py 1023\n",
      "The caller function is '_functional_construction_call' in module 'keras.engine.base_layer' (current in base_layer.py 2358)\n",
      "The caller function is '<listcomp>' in module 'tensorflow.python.util.nest' (currently in keras_tensor.py 670)\n",
      "----------------------------debug keras_tensor.py 243\n",
      "The caller function is 'keras_tensor_to_placeholder' in module 'keras.engine.keras_tensor' (currently in keras_tensor.py 248)\n",
      "keras_tensor.py self: KerasTensor(type_spec=TensorSpec(shape=(None, 14, 14, 512), dtype=tf.float32, name=None), name='block5_conv1/Relu:0', description=\"created by layer 'block5_conv1'\")\n",
      "component_to_placeholder: <function KerasTensor._to_placeholder.<locals>.component_to_placeholder at 0x0000013B7E2BB9A0>\n",
      "self.type_spec: TensorSpec(shape=(None, 14, 14, 512), dtype=tf.float32, name=None)\n",
      "The caller function is '<listcomp>' in module 'tensorflow.python.util.nest' (current in keras.engine.keras_tensor 238)\n",
      "The caller function is 'placeholder' in module 'tensorflow.python.ops.array_ops' (current in gen_array_ops.py 6902)\n",
      "name: None\n",
      "dtype: 1\n",
      "The caller function is '<listcomp>' in module 'tensorflow.python.util.nest' (currently in keras_tensor.py 670)\n",
      "----------------------------debug base_layer.py 2441\n",
      "----------------------------debug traceback_utils.py 101\n",
      "The caller function is '_infer_output_signature' in module 'keras.engine.base_layer' (current in base_layer.py 989)\n",
      "----------------------------debug base_layer.py 2462\n",
      "----------------------------debug base_layer.py 2471\n",
      "----------------------------debug base_layer.py 983\n",
      "str(self.name): block5_conv3\n",
      "The caller function is 'error_handler' in module 'keras.utils.traceback_utils' (current in base_layer.py 989)\n",
      "args: (<KerasTensor: shape=(None, 14, 14, 512) dtype=float32 (created by layer 'block5_conv2')>,)\n",
      "kwargs: {}\n",
      "----------------------------debug base_layer.py 1009\n",
      "----------------------------debug base_layer.py 1011\n",
      "----------------------------debug base_layer.py 1013\n",
      "----------------------------debug base_layer.py 1023\n",
      "The caller function is '_functional_construction_call' in module 'keras.engine.base_layer' (current in base_layer.py 2358)\n",
      "The caller function is '<listcomp>' in module 'tensorflow.python.util.nest' (currently in keras_tensor.py 670)\n",
      "----------------------------debug keras_tensor.py 243\n",
      "The caller function is 'keras_tensor_to_placeholder' in module 'keras.engine.keras_tensor' (currently in keras_tensor.py 248)\n",
      "keras_tensor.py self: KerasTensor(type_spec=TensorSpec(shape=(None, 14, 14, 512), dtype=tf.float32, name=None), name='block5_conv2/Relu:0', description=\"created by layer 'block5_conv2'\")\n",
      "component_to_placeholder: <function KerasTensor._to_placeholder.<locals>.component_to_placeholder at 0x0000013B7E2BBA30>\n",
      "self.type_spec: TensorSpec(shape=(None, 14, 14, 512), dtype=tf.float32, name=None)\n",
      "The caller function is '<listcomp>' in module 'tensorflow.python.util.nest' (current in keras.engine.keras_tensor 238)\n",
      "The caller function is 'placeholder' in module 'tensorflow.python.ops.array_ops' (current in gen_array_ops.py 6902)\n",
      "name: None\n",
      "dtype: 1\n",
      "The caller function is '<listcomp>' in module 'tensorflow.python.util.nest' (currently in keras_tensor.py 670)\n",
      "----------------------------debug base_layer.py 2441\n",
      "----------------------------debug traceback_utils.py 101\n",
      "The caller function is '_infer_output_signature' in module 'keras.engine.base_layer' (current in base_layer.py 989)\n",
      "----------------------------debug base_layer.py 2462\n",
      "----------------------------debug base_layer.py 2471\n",
      "----------------------------debug base_layer.py 983\n",
      "str(self.name): block5_conv4\n",
      "The caller function is 'error_handler' in module 'keras.utils.traceback_utils' (current in base_layer.py 989)\n",
      "args: (<KerasTensor: shape=(None, 14, 14, 512) dtype=float32 (created by layer 'block5_conv3')>,)\n",
      "kwargs: {}\n",
      "----------------------------debug base_layer.py 1009\n",
      "----------------------------debug base_layer.py 1011\n",
      "----------------------------debug base_layer.py 1013\n",
      "----------------------------debug base_layer.py 1023\n",
      "The caller function is '_functional_construction_call' in module 'keras.engine.base_layer' (current in base_layer.py 2358)\n",
      "The caller function is '<listcomp>' in module 'tensorflow.python.util.nest' (currently in keras_tensor.py 670)\n",
      "----------------------------debug keras_tensor.py 243\n",
      "The caller function is 'keras_tensor_to_placeholder' in module 'keras.engine.keras_tensor' (currently in keras_tensor.py 248)\n",
      "keras_tensor.py self: KerasTensor(type_spec=TensorSpec(shape=(None, 14, 14, 512), dtype=tf.float32, name=None), name='block5_conv3/Relu:0', description=\"created by layer 'block5_conv3'\")\n",
      "component_to_placeholder: <function KerasTensor._to_placeholder.<locals>.component_to_placeholder at 0x0000013B7E2BB9A0>\n",
      "self.type_spec: TensorSpec(shape=(None, 14, 14, 512), dtype=tf.float32, name=None)\n",
      "The caller function is '<listcomp>' in module 'tensorflow.python.util.nest' (current in keras.engine.keras_tensor 238)\n",
      "The caller function is 'placeholder' in module 'tensorflow.python.ops.array_ops' (current in gen_array_ops.py 6902)\n",
      "name: None\n",
      "dtype: 1\n",
      "The caller function is '<listcomp>' in module 'tensorflow.python.util.nest' (currently in keras_tensor.py 670)\n",
      "----------------------------debug base_layer.py 2441\n",
      "----------------------------debug traceback_utils.py 101\n",
      "The caller function is '_infer_output_signature' in module 'keras.engine.base_layer' (current in base_layer.py 989)\n",
      "----------------------------debug base_layer.py 2462\n",
      "----------------------------debug base_layer.py 2471\n",
      "----------------------------debug base_layer.py 983\n",
      "str(self.name): block5_pool\n",
      "The caller function is 'error_handler' in module 'keras.utils.traceback_utils' (current in base_layer.py 989)\n",
      "args: (<KerasTensor: shape=(None, 14, 14, 512) dtype=float32 (created by layer 'block5_conv4')>,)\n",
      "kwargs: {}\n",
      "----------------------------debug base_layer.py 1009\n",
      "----------------------------debug base_layer.py 1011\n",
      "----------------------------debug base_layer.py 1013\n",
      "----------------------------debug base_layer.py 1023\n",
      "The caller function is '_functional_construction_call' in module 'keras.engine.base_layer' (current in base_layer.py 2358)\n",
      "The caller function is '<listcomp>' in module 'tensorflow.python.util.nest' (currently in keras_tensor.py 670)\n",
      "----------------------------debug keras_tensor.py 243\n",
      "The caller function is 'keras_tensor_to_placeholder' in module 'keras.engine.keras_tensor' (currently in keras_tensor.py 248)\n",
      "keras_tensor.py self: KerasTensor(type_spec=TensorSpec(shape=(None, 14, 14, 512), dtype=tf.float32, name=None), name='block5_conv4/Relu:0', description=\"created by layer 'block5_conv4'\")\n",
      "component_to_placeholder: <function KerasTensor._to_placeholder.<locals>.component_to_placeholder at 0x0000013B7E2BBA30>\n",
      "self.type_spec: TensorSpec(shape=(None, 14, 14, 512), dtype=tf.float32, name=None)\n",
      "The caller function is '<listcomp>' in module 'tensorflow.python.util.nest' (current in keras.engine.keras_tensor 238)\n",
      "The caller function is 'placeholder' in module 'tensorflow.python.ops.array_ops' (current in gen_array_ops.py 6902)\n",
      "name: None\n",
      "dtype: 1\n",
      "The caller function is '<listcomp>' in module 'tensorflow.python.util.nest' (currently in keras_tensor.py 670)\n",
      "----------------------------debug base_layer.py 2441\n",
      "----------------------------debug traceback_utils.py 101\n",
      "The caller function is '_infer_output_signature' in module 'keras.engine.base_layer' (current in base_layer.py 989)\n",
      "----------------------------debug base_layer.py 2462\n",
      "----------------------------debug base_layer.py 2471\n",
      "----------------------------debug functional.py 1102\n",
      "inputs: [<KerasTensor: shape=(None, 224, 224, 3) dtype=float32 (created by layer 'input_1')>]\n",
      "The caller function is '_init_graph_network' in module 'keras.engine.functional' (current in functional.py 1108)\n",
      "----------------------------debug functional.py 1159\n",
      "len nodes_by_depth: 22\n",
      "layers_with_complete_input: ['block1_conv1', 'block1_conv2', 'block1_pool', 'block2_conv1', 'block2_conv2', 'block2_pool', 'block3_conv1', 'block3_conv2', 'block3_conv3', 'block3_conv4', 'block3_pool', 'block4_conv1', 'block4_conv2', 'block4_conv3', 'block4_conv4', 'block4_pool', 'block5_conv1', 'block5_conv2', 'block5_conv3', 'block5_conv4', 'block5_pool']\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "----------------------------debug base_layer.py 983\n",
      "str(self.name): vgg19\n",
      "The caller function is 'error_handler' in module 'keras.utils.traceback_utils' (current in base_layer.py 989)\n",
      "args: (<KerasTensor: shape=(None, 224, 224, 3) dtype=float32 (created by layer 'input_2')>,)\n",
      "kwargs: {}\n",
      "----------------------------debug base_layer.py 1009\n",
      "----------------------------debug base_layer.py 1011\n",
      "----------------------------debug base_layer.py 1013\n",
      "----------------------------debug base_layer.py 1023\n",
      "The caller function is '_functional_construction_call' in module 'keras.engine.base_layer' (current in base_layer.py 2358)\n",
      "The caller function is '<listcomp>' in module 'tensorflow.python.util.nest' (currently in keras_tensor.py 670)\n",
      "----------------------------debug keras_tensor.py 243\n",
      "The caller function is 'keras_tensor_to_placeholder' in module 'keras.engine.keras_tensor' (currently in keras_tensor.py 248)\n",
      "keras_tensor.py self: KerasTensor(type_spec=TensorSpec(shape=(None, 224, 224, 3), dtype=tf.float32, name='input_2'), name='input_2', description=\"created by layer 'input_2'\")\n",
      "component_to_placeholder: <function KerasTensor._to_placeholder.<locals>.component_to_placeholder at 0x0000013B7E2BBAC0>\n",
      "self.type_spec: TensorSpec(shape=(None, 224, 224, 3), dtype=tf.float32, name='input_2')\n",
      "The caller function is '<listcomp>' in module 'tensorflow.python.util.nest' (current in keras.engine.keras_tensor 238)\n",
      "The caller function is 'placeholder' in module 'tensorflow.python.ops.array_ops' (current in gen_array_ops.py 6902)\n",
      "name: None\n",
      "dtype: 1\n",
      "The caller function is '<listcomp>' in module 'tensorflow.python.util.nest' (currently in keras_tensor.py 670)\n",
      "The caller function is '<listcomp>' in module 'tensorflow.python.util.nest' (currently in keras_tensor.py 670)\n",
      "----------------------------debug base_layer.py 2441\n",
      "----------------------------debug traceback_utils.py 101\n",
      "The caller function is '_infer_output_signature' in module 'keras.engine.base_layer' (current in base_layer.py 989)\n",
      "----------------------------debug base_layer.py 983\n",
      "str(self.name): block1_conv1\n",
      "The caller function is 'error_handler' in module 'keras.utils.traceback_utils' (current in base_layer.py 989)\n",
      "args: (<tf.Tensor 'Placeholder:0' shape=(None, 224, 224, 3) dtype=float32>,)\n",
      "kwargs: {}\n",
      "----------------------------debug base_layer.py 1009\n",
      "----------------------------debug base_layer.py 1011\n",
      "----------------------------debug base_layer.py 1013\n",
      "----------------------------debug base_layer.py 1048\n",
      "----------------------------debug base_layer.py 1112\n",
      "inputs: Tensor(\"Placeholder:0\", shape=(None, 224, 224, 3), dtype=float32)\n",
      "----------------------------debug traceback_utils.py 101\n",
      "The caller function is '__call__' in module 'keras.engine.base_layer' (current in base_layer.py 989)\n",
      "outputs: Tensor(\"vgg19/block1_conv1/Relu:0\", shape=(None, 224, 224, 64), dtype=float32)\n",
      "----------------------------debug base_layer.py 983\n",
      "str(self.name): block1_conv2\n",
      "The caller function is 'error_handler' in module 'keras.utils.traceback_utils' (current in base_layer.py 989)\n",
      "args: (<tf.Tensor 'vgg19/block1_conv1/Relu:0' shape=(None, 224, 224, 64) dtype=float32>,)\n",
      "kwargs: {}\n",
      "----------------------------debug base_layer.py 1009\n",
      "----------------------------debug base_layer.py 1011\n",
      "----------------------------debug base_layer.py 1013\n",
      "----------------------------debug base_layer.py 1048\n",
      "----------------------------debug base_layer.py 1112\n",
      "inputs: Tensor(\"vgg19/block1_conv1/Relu:0\", shape=(None, 224, 224, 64), dtype=float32)\n",
      "----------------------------debug traceback_utils.py 101\n",
      "The caller function is '__call__' in module 'keras.engine.base_layer' (current in base_layer.py 989)\n",
      "outputs: Tensor(\"vgg19/block1_conv2/Relu:0\", shape=(None, 224, 224, 64), dtype=float32)\n",
      "----------------------------debug base_layer.py 983\n",
      "str(self.name): block1_pool\n",
      "The caller function is 'error_handler' in module 'keras.utils.traceback_utils' (current in base_layer.py 989)\n",
      "args: (<tf.Tensor 'vgg19/block1_conv2/Relu:0' shape=(None, 224, 224, 64) dtype=float32>,)\n",
      "kwargs: {}\n",
      "----------------------------debug base_layer.py 1009\n",
      "----------------------------debug base_layer.py 1011\n",
      "----------------------------debug base_layer.py 1013\n",
      "----------------------------debug base_layer.py 1048\n",
      "----------------------------debug base_layer.py 1112\n",
      "inputs: Tensor(\"vgg19/block1_conv2/Relu:0\", shape=(None, 224, 224, 64), dtype=float32)\n",
      "----------------------------debug traceback_utils.py 101\n",
      "The caller function is '__call__' in module 'keras.engine.base_layer' (current in base_layer.py 989)\n",
      "outputs: Tensor(\"vgg19/block1_pool/MaxPool:0\", shape=(None, 112, 112, 64), dtype=float32)\n",
      "----------------------------debug base_layer.py 983\n",
      "str(self.name): block2_conv1\n",
      "The caller function is 'error_handler' in module 'keras.utils.traceback_utils' (current in base_layer.py 989)\n",
      "args: (<tf.Tensor 'vgg19/block1_pool/MaxPool:0' shape=(None, 112, 112, 64) dtype=float32>,)\n",
      "kwargs: {}\n",
      "----------------------------debug base_layer.py 1009\n",
      "----------------------------debug base_layer.py 1011\n",
      "----------------------------debug base_layer.py 1013\n",
      "----------------------------debug base_layer.py 1048\n",
      "----------------------------debug base_layer.py 1112\n",
      "inputs: Tensor(\"vgg19/block1_pool/MaxPool:0\", shape=(None, 112, 112, 64), dtype=float32)\n",
      "----------------------------debug traceback_utils.py 101\n",
      "The caller function is '__call__' in module 'keras.engine.base_layer' (current in base_layer.py 989)\n",
      "outputs: Tensor(\"vgg19/block2_conv1/Relu:0\", shape=(None, 112, 112, 128), dtype=float32)\n",
      "----------------------------debug base_layer.py 983\n",
      "str(self.name): block2_conv2\n",
      "The caller function is 'error_handler' in module 'keras.utils.traceback_utils' (current in base_layer.py 989)\n",
      "args: (<tf.Tensor 'vgg19/block2_conv1/Relu:0' shape=(None, 112, 112, 128) dtype=float32>,)\n",
      "kwargs: {}\n",
      "----------------------------debug base_layer.py 1009\n",
      "----------------------------debug base_layer.py 1011\n",
      "----------------------------debug base_layer.py 1013\n",
      "----------------------------debug base_layer.py 1048\n",
      "----------------------------debug base_layer.py 1112\n",
      "inputs: Tensor(\"vgg19/block2_conv1/Relu:0\", shape=(None, 112, 112, 128), dtype=float32)\n",
      "----------------------------debug traceback_utils.py 101\n",
      "The caller function is '__call__' in module 'keras.engine.base_layer' (current in base_layer.py 989)\n",
      "outputs: Tensor(\"vgg19/block2_conv2/Relu:0\", shape=(None, 112, 112, 128), dtype=float32)\n",
      "----------------------------debug base_layer.py 983\n",
      "str(self.name): block2_pool\n",
      "The caller function is 'error_handler' in module 'keras.utils.traceback_utils' (current in base_layer.py 989)\n",
      "args: (<tf.Tensor 'vgg19/block2_conv2/Relu:0' shape=(None, 112, 112, 128) dtype=float32>,)\n",
      "kwargs: {}\n",
      "----------------------------debug base_layer.py 1009\n",
      "----------------------------debug base_layer.py 1011\n",
      "----------------------------debug base_layer.py 1013\n",
      "----------------------------debug base_layer.py 1048\n",
      "----------------------------debug base_layer.py 1112\n",
      "inputs: Tensor(\"vgg19/block2_conv2/Relu:0\", shape=(None, 112, 112, 128), dtype=float32)\n",
      "----------------------------debug traceback_utils.py 101\n",
      "The caller function is '__call__' in module 'keras.engine.base_layer' (current in base_layer.py 989)\n",
      "outputs: Tensor(\"vgg19/block2_pool/MaxPool:0\", shape=(None, 56, 56, 128), dtype=float32)\n",
      "----------------------------debug base_layer.py 983\n",
      "str(self.name): block3_conv1\n",
      "The caller function is 'error_handler' in module 'keras.utils.traceback_utils' (current in base_layer.py 989)\n",
      "args: (<tf.Tensor 'vgg19/block2_pool/MaxPool:0' shape=(None, 56, 56, 128) dtype=float32>,)\n",
      "kwargs: {}\n",
      "----------------------------debug base_layer.py 1009\n",
      "----------------------------debug base_layer.py 1011\n",
      "----------------------------debug base_layer.py 1013\n",
      "----------------------------debug base_layer.py 1048\n",
      "----------------------------debug base_layer.py 1112\n",
      "inputs: Tensor(\"vgg19/block2_pool/MaxPool:0\", shape=(None, 56, 56, 128), dtype=float32)\n",
      "----------------------------debug traceback_utils.py 101\n",
      "The caller function is '__call__' in module 'keras.engine.base_layer' (current in base_layer.py 989)\n",
      "outputs: Tensor(\"vgg19/block3_conv1/Relu:0\", shape=(None, 56, 56, 256), dtype=float32)\n",
      "----------------------------debug base_layer.py 983\n",
      "str(self.name): block3_conv2\n",
      "The caller function is 'error_handler' in module 'keras.utils.traceback_utils' (current in base_layer.py 989)\n",
      "args: (<tf.Tensor 'vgg19/block3_conv1/Relu:0' shape=(None, 56, 56, 256) dtype=float32>,)\n",
      "kwargs: {}\n",
      "----------------------------debug base_layer.py 1009\n",
      "----------------------------debug base_layer.py 1011\n",
      "----------------------------debug base_layer.py 1013\n",
      "----------------------------debug base_layer.py 1048\n",
      "----------------------------debug base_layer.py 1112\n",
      "inputs: Tensor(\"vgg19/block3_conv1/Relu:0\", shape=(None, 56, 56, 256), dtype=float32)\n",
      "----------------------------debug traceback_utils.py 101\n",
      "The caller function is '__call__' in module 'keras.engine.base_layer' (current in base_layer.py 989)\n",
      "outputs: Tensor(\"vgg19/block3_conv2/Relu:0\", shape=(None, 56, 56, 256), dtype=float32)\n",
      "----------------------------debug base_layer.py 983\n",
      "str(self.name): block3_conv3\n",
      "The caller function is 'error_handler' in module 'keras.utils.traceback_utils' (current in base_layer.py 989)\n",
      "args: (<tf.Tensor 'vgg19/block3_conv2/Relu:0' shape=(None, 56, 56, 256) dtype=float32>,)\n",
      "kwargs: {}\n",
      "----------------------------debug base_layer.py 1009\n",
      "----------------------------debug base_layer.py 1011\n",
      "----------------------------debug base_layer.py 1013\n",
      "----------------------------debug base_layer.py 1048\n",
      "----------------------------debug base_layer.py 1112\n",
      "inputs: Tensor(\"vgg19/block3_conv2/Relu:0\", shape=(None, 56, 56, 256), dtype=float32)\n",
      "----------------------------debug traceback_utils.py 101\n",
      "The caller function is '__call__' in module 'keras.engine.base_layer' (current in base_layer.py 989)\n",
      "outputs: Tensor(\"vgg19/block3_conv3/Relu:0\", shape=(None, 56, 56, 256), dtype=float32)\n",
      "----------------------------debug base_layer.py 983\n",
      "str(self.name): block3_conv4\n",
      "The caller function is 'error_handler' in module 'keras.utils.traceback_utils' (current in base_layer.py 989)\n",
      "args: (<tf.Tensor 'vgg19/block3_conv3/Relu:0' shape=(None, 56, 56, 256) dtype=float32>,)\n",
      "kwargs: {}\n",
      "----------------------------debug base_layer.py 1009\n",
      "----------------------------debug base_layer.py 1011\n",
      "----------------------------debug base_layer.py 1013\n",
      "----------------------------debug base_layer.py 1048\n",
      "----------------------------debug base_layer.py 1112\n",
      "inputs: Tensor(\"vgg19/block3_conv3/Relu:0\", shape=(None, 56, 56, 256), dtype=float32)\n",
      "----------------------------debug traceback_utils.py 101\n",
      "The caller function is '__call__' in module 'keras.engine.base_layer' (current in base_layer.py 989)\n",
      "outputs: Tensor(\"vgg19/block3_conv4/Relu:0\", shape=(None, 56, 56, 256), dtype=float32)\n",
      "----------------------------debug base_layer.py 983\n",
      "str(self.name): block3_pool\n",
      "The caller function is 'error_handler' in module 'keras.utils.traceback_utils' (current in base_layer.py 989)\n",
      "args: (<tf.Tensor 'vgg19/block3_conv4/Relu:0' shape=(None, 56, 56, 256) dtype=float32>,)\n",
      "kwargs: {}\n",
      "----------------------------debug base_layer.py 1009\n",
      "----------------------------debug base_layer.py 1011\n",
      "----------------------------debug base_layer.py 1013\n",
      "----------------------------debug base_layer.py 1048\n",
      "----------------------------debug base_layer.py 1112\n",
      "inputs: Tensor(\"vgg19/block3_conv4/Relu:0\", shape=(None, 56, 56, 256), dtype=float32)\n",
      "----------------------------debug traceback_utils.py 101\n",
      "The caller function is '__call__' in module 'keras.engine.base_layer' (current in base_layer.py 989)\n",
      "outputs: Tensor(\"vgg19/block3_pool/MaxPool:0\", shape=(None, 28, 28, 256), dtype=float32)\n",
      "----------------------------debug base_layer.py 983\n",
      "str(self.name): block4_conv1\n",
      "The caller function is 'error_handler' in module 'keras.utils.traceback_utils' (current in base_layer.py 989)\n",
      "args: (<tf.Tensor 'vgg19/block3_pool/MaxPool:0' shape=(None, 28, 28, 256) dtype=float32>,)\n",
      "kwargs: {}\n",
      "----------------------------debug base_layer.py 1009\n",
      "----------------------------debug base_layer.py 1011\n",
      "----------------------------debug base_layer.py 1013\n",
      "----------------------------debug base_layer.py 1048\n",
      "----------------------------debug base_layer.py 1112\n",
      "inputs: Tensor(\"vgg19/block3_pool/MaxPool:0\", shape=(None, 28, 28, 256), dtype=float32)\n",
      "----------------------------debug traceback_utils.py 101\n",
      "The caller function is '__call__' in module 'keras.engine.base_layer' (current in base_layer.py 989)\n",
      "outputs: Tensor(\"vgg19/block4_conv1/Relu:0\", shape=(None, 28, 28, 512), dtype=float32)\n",
      "----------------------------debug base_layer.py 983\n",
      "str(self.name): block4_conv2\n",
      "The caller function is 'error_handler' in module 'keras.utils.traceback_utils' (current in base_layer.py 989)\n",
      "args: (<tf.Tensor 'vgg19/block4_conv1/Relu:0' shape=(None, 28, 28, 512) dtype=float32>,)\n",
      "kwargs: {}\n",
      "----------------------------debug base_layer.py 1009\n",
      "----------------------------debug base_layer.py 1011\n",
      "----------------------------debug base_layer.py 1013\n",
      "----------------------------debug base_layer.py 1048\n",
      "----------------------------debug base_layer.py 1112\n",
      "inputs: Tensor(\"vgg19/block4_conv1/Relu:0\", shape=(None, 28, 28, 512), dtype=float32)\n",
      "----------------------------debug traceback_utils.py 101\n",
      "The caller function is '__call__' in module 'keras.engine.base_layer' (current in base_layer.py 989)\n",
      "outputs: Tensor(\"vgg19/block4_conv2/Relu:0\", shape=(None, 28, 28, 512), dtype=float32)\n",
      "----------------------------debug base_layer.py 983\n",
      "str(self.name): block4_conv3\n",
      "The caller function is 'error_handler' in module 'keras.utils.traceback_utils' (current in base_layer.py 989)\n",
      "args: (<tf.Tensor 'vgg19/block4_conv2/Relu:0' shape=(None, 28, 28, 512) dtype=float32>,)\n",
      "kwargs: {}\n",
      "----------------------------debug base_layer.py 1009\n",
      "----------------------------debug base_layer.py 1011\n",
      "----------------------------debug base_layer.py 1013\n",
      "----------------------------debug base_layer.py 1048\n",
      "----------------------------debug base_layer.py 1112\n",
      "inputs: Tensor(\"vgg19/block4_conv2/Relu:0\", shape=(None, 28, 28, 512), dtype=float32)\n",
      "----------------------------debug traceback_utils.py 101\n",
      "The caller function is '__call__' in module 'keras.engine.base_layer' (current in base_layer.py 989)\n",
      "outputs: Tensor(\"vgg19/block4_conv3/Relu:0\", shape=(None, 28, 28, 512), dtype=float32)\n",
      "----------------------------debug base_layer.py 983\n",
      "str(self.name): block4_conv4\n",
      "The caller function is 'error_handler' in module 'keras.utils.traceback_utils' (current in base_layer.py 989)\n",
      "args: (<tf.Tensor 'vgg19/block4_conv3/Relu:0' shape=(None, 28, 28, 512) dtype=float32>,)\n",
      "kwargs: {}\n",
      "----------------------------debug base_layer.py 1009\n",
      "----------------------------debug base_layer.py 1011\n",
      "----------------------------debug base_layer.py 1013\n",
      "----------------------------debug base_layer.py 1048\n",
      "----------------------------debug base_layer.py 1112\n",
      "inputs: Tensor(\"vgg19/block4_conv3/Relu:0\", shape=(None, 28, 28, 512), dtype=float32)\n",
      "----------------------------debug traceback_utils.py 101\n",
      "The caller function is '__call__' in module 'keras.engine.base_layer' (current in base_layer.py 989)\n",
      "outputs: Tensor(\"vgg19/block4_conv4/Relu:0\", shape=(None, 28, 28, 512), dtype=float32)\n",
      "----------------------------debug base_layer.py 983\n",
      "str(self.name): block4_pool\n",
      "The caller function is 'error_handler' in module 'keras.utils.traceback_utils' (current in base_layer.py 989)\n",
      "args: (<tf.Tensor 'vgg19/block4_conv4/Relu:0' shape=(None, 28, 28, 512) dtype=float32>,)\n",
      "kwargs: {}\n",
      "----------------------------debug base_layer.py 1009\n",
      "----------------------------debug base_layer.py 1011\n",
      "----------------------------debug base_layer.py 1013\n",
      "----------------------------debug base_layer.py 1048\n",
      "----------------------------debug base_layer.py 1112\n",
      "inputs: Tensor(\"vgg19/block4_conv4/Relu:0\", shape=(None, 28, 28, 512), dtype=float32)\n",
      "----------------------------debug traceback_utils.py 101\n",
      "The caller function is '__call__' in module 'keras.engine.base_layer' (current in base_layer.py 989)\n",
      "outputs: Tensor(\"vgg19/block4_pool/MaxPool:0\", shape=(None, 14, 14, 512), dtype=float32)\n",
      "----------------------------debug base_layer.py 983\n",
      "str(self.name): block5_conv1\n",
      "The caller function is 'error_handler' in module 'keras.utils.traceback_utils' (current in base_layer.py 989)\n",
      "args: (<tf.Tensor 'vgg19/block4_pool/MaxPool:0' shape=(None, 14, 14, 512) dtype=float32>,)\n",
      "kwargs: {}\n",
      "----------------------------debug base_layer.py 1009\n",
      "----------------------------debug base_layer.py 1011\n",
      "----------------------------debug base_layer.py 1013\n",
      "----------------------------debug base_layer.py 1048\n",
      "----------------------------debug base_layer.py 1112\n",
      "inputs: Tensor(\"vgg19/block4_pool/MaxPool:0\", shape=(None, 14, 14, 512), dtype=float32)\n",
      "----------------------------debug traceback_utils.py 101\n",
      "The caller function is '__call__' in module 'keras.engine.base_layer' (current in base_layer.py 989)\n",
      "outputs: Tensor(\"vgg19/block5_conv1/Relu:0\", shape=(None, 14, 14, 512), dtype=float32)\n",
      "----------------------------debug base_layer.py 983\n",
      "str(self.name): block5_conv2\n",
      "The caller function is 'error_handler' in module 'keras.utils.traceback_utils' (current in base_layer.py 989)\n",
      "args: (<tf.Tensor 'vgg19/block5_conv1/Relu:0' shape=(None, 14, 14, 512) dtype=float32>,)\n",
      "kwargs: {}\n",
      "----------------------------debug base_layer.py 1009\n",
      "----------------------------debug base_layer.py 1011\n",
      "----------------------------debug base_layer.py 1013\n",
      "----------------------------debug base_layer.py 1048\n",
      "----------------------------debug base_layer.py 1112\n",
      "inputs: Tensor(\"vgg19/block5_conv1/Relu:0\", shape=(None, 14, 14, 512), dtype=float32)\n",
      "----------------------------debug traceback_utils.py 101\n",
      "The caller function is '__call__' in module 'keras.engine.base_layer' (current in base_layer.py 989)\n",
      "outputs: Tensor(\"vgg19/block5_conv2/Relu:0\", shape=(None, 14, 14, 512), dtype=float32)\n",
      "----------------------------debug base_layer.py 983\n",
      "str(self.name): block5_conv3\n",
      "The caller function is 'error_handler' in module 'keras.utils.traceback_utils' (current in base_layer.py 989)\n",
      "args: (<tf.Tensor 'vgg19/block5_conv2/Relu:0' shape=(None, 14, 14, 512) dtype=float32>,)\n",
      "kwargs: {}\n",
      "----------------------------debug base_layer.py 1009\n",
      "----------------------------debug base_layer.py 1011\n",
      "----------------------------debug base_layer.py 1013\n",
      "----------------------------debug base_layer.py 1048\n",
      "----------------------------debug base_layer.py 1112\n",
      "inputs: Tensor(\"vgg19/block5_conv2/Relu:0\", shape=(None, 14, 14, 512), dtype=float32)\n",
      "----------------------------debug traceback_utils.py 101\n",
      "The caller function is '__call__' in module 'keras.engine.base_layer' (current in base_layer.py 989)\n",
      "outputs: Tensor(\"vgg19/block5_conv3/Relu:0\", shape=(None, 14, 14, 512), dtype=float32)\n",
      "----------------------------debug base_layer.py 983\n",
      "str(self.name): block5_conv4\n",
      "The caller function is 'error_handler' in module 'keras.utils.traceback_utils' (current in base_layer.py 989)\n",
      "args: (<tf.Tensor 'vgg19/block5_conv3/Relu:0' shape=(None, 14, 14, 512) dtype=float32>,)\n",
      "kwargs: {}\n",
      "----------------------------debug base_layer.py 1009\n",
      "----------------------------debug base_layer.py 1011\n",
      "----------------------------debug base_layer.py 1013\n",
      "----------------------------debug base_layer.py 1048\n",
      "----------------------------debug base_layer.py 1112\n",
      "inputs: Tensor(\"vgg19/block5_conv3/Relu:0\", shape=(None, 14, 14, 512), dtype=float32)\n",
      "----------------------------debug traceback_utils.py 101\n",
      "The caller function is '__call__' in module 'keras.engine.base_layer' (current in base_layer.py 989)\n",
      "outputs: Tensor(\"vgg19/block5_conv4/Relu:0\", shape=(None, 14, 14, 512), dtype=float32)\n",
      "----------------------------debug base_layer.py 983\n",
      "str(self.name): block5_pool\n",
      "The caller function is 'error_handler' in module 'keras.utils.traceback_utils' (current in base_layer.py 989)\n",
      "args: (<tf.Tensor 'vgg19/block5_conv4/Relu:0' shape=(None, 14, 14, 512) dtype=float32>,)\n",
      "kwargs: {}\n",
      "----------------------------debug base_layer.py 1009\n",
      "----------------------------debug base_layer.py 1011\n",
      "----------------------------debug base_layer.py 1013\n",
      "----------------------------debug base_layer.py 1048\n",
      "----------------------------debug base_layer.py 1112\n",
      "inputs: Tensor(\"vgg19/block5_conv4/Relu:0\", shape=(None, 14, 14, 512), dtype=float32)\n",
      "----------------------------debug traceback_utils.py 101\n",
      "The caller function is '__call__' in module 'keras.engine.base_layer' (current in base_layer.py 989)\n",
      "outputs: Tensor(\"vgg19/block5_pool/MaxPool:0\", shape=(None, 7, 7, 512), dtype=float32)\n",
      "----------------------------debug base_layer.py 2462\n",
      "----------------------------debug base_layer.py 2471\n",
      "----------------------------debug base_layer.py 983\n",
      "str(self.name): flatten\n",
      "The caller function is 'error_handler' in module 'keras.utils.traceback_utils' (current in base_layer.py 989)\n",
      "args: (<KerasTensor: shape=(None, 7, 7, 512) dtype=float32 (created by layer 'vgg19')>,)\n",
      "kwargs: {}\n",
      "----------------------------debug base_layer.py 1009\n",
      "----------------------------debug base_layer.py 1011\n",
      "----------------------------debug base_layer.py 1013\n",
      "----------------------------debug base_layer.py 1023\n",
      "The caller function is '_functional_construction_call' in module 'keras.engine.base_layer' (current in base_layer.py 2358)\n",
      "The caller function is '<listcomp>' in module 'tensorflow.python.util.nest' (currently in keras_tensor.py 670)\n",
      "----------------------------debug keras_tensor.py 243\n",
      "The caller function is 'keras_tensor_to_placeholder' in module 'keras.engine.keras_tensor' (currently in keras_tensor.py 248)\n",
      "keras_tensor.py self: KerasTensor(type_spec=TensorSpec(shape=(None, 7, 7, 512), dtype=tf.float32, name=None), name='vgg19/block5_pool/MaxPool:0', description=\"created by layer 'vgg19'\")\n",
      "component_to_placeholder: <function KerasTensor._to_placeholder.<locals>.component_to_placeholder at 0x0000013B7E2BBB50>\n",
      "self.type_spec: TensorSpec(shape=(None, 7, 7, 512), dtype=tf.float32, name=None)\n",
      "The caller function is '<listcomp>' in module 'tensorflow.python.util.nest' (current in keras.engine.keras_tensor 238)\n",
      "The caller function is 'placeholder' in module 'tensorflow.python.ops.array_ops' (current in gen_array_ops.py 6902)\n",
      "name: None\n",
      "dtype: 1\n",
      "The caller function is '<listcomp>' in module 'tensorflow.python.util.nest' (currently in keras_tensor.py 670)\n",
      "----------------------------debug base_layer.py 2441\n",
      "----------------------------debug traceback_utils.py 101\n",
      "The caller function is '_infer_output_signature' in module 'keras.engine.base_layer' (current in base_layer.py 989)\n",
      "----------------------------debug base_layer.py 2462\n",
      "----------------------------debug base_layer.py 2471\n",
      "----------------------------debug base_layer.py 983\n",
      "str(self.name): dense\n",
      "The caller function is 'error_handler' in module 'keras.utils.traceback_utils' (current in base_layer.py 989)\n",
      "args: (<KerasTensor: shape=(None, 25088) dtype=float32 (created by layer 'flatten')>,)\n",
      "kwargs: {}\n",
      "----------------------------debug base_layer.py 1009\n",
      "----------------------------debug base_layer.py 1011\n",
      "----------------------------debug base_layer.py 1013\n",
      "----------------------------debug base_layer.py 1023\n",
      "The caller function is '_functional_construction_call' in module 'keras.engine.base_layer' (current in base_layer.py 2358)\n",
      "The caller function is '<listcomp>' in module 'tensorflow.python.util.nest' (currently in keras_tensor.py 670)\n",
      "----------------------------debug keras_tensor.py 243\n",
      "The caller function is 'keras_tensor_to_placeholder' in module 'keras.engine.keras_tensor' (currently in keras_tensor.py 248)\n",
      "keras_tensor.py self: KerasTensor(type_spec=TensorSpec(shape=(None, 25088), dtype=tf.float32, name=None), name='flatten/Reshape:0', description=\"created by layer 'flatten'\")\n",
      "component_to_placeholder: <function KerasTensor._to_placeholder.<locals>.component_to_placeholder at 0x0000013B7E2BBD00>\n",
      "self.type_spec: TensorSpec(shape=(None, 25088), dtype=tf.float32, name=None)\n",
      "The caller function is '<listcomp>' in module 'tensorflow.python.util.nest' (current in keras.engine.keras_tensor 238)\n",
      "The caller function is 'placeholder' in module 'tensorflow.python.ops.array_ops' (current in gen_array_ops.py 6902)\n",
      "name: None\n",
      "dtype: 1\n",
      "The caller function is '<listcomp>' in module 'tensorflow.python.util.nest' (currently in keras_tensor.py 670)\n",
      "----------------------------debug base_layer.py 2441\n",
      "----------------------------debug traceback_utils.py 101\n",
      "The caller function is '_infer_output_signature' in module 'keras.engine.base_layer' (current in base_layer.py 989)\n",
      "----------------------------debug base_layer.py 2462\n",
      "----------------------------debug base_layer.py 2471\n",
      "----------------------------debug base_layer.py 983\n",
      "str(self.name): leaky_re_lu\n",
      "The caller function is 'error_handler' in module 'keras.utils.traceback_utils' (current in base_layer.py 989)\n",
      "args: (<KerasTensor: shape=(None, 128) dtype=float32 (created by layer 'dense')>,)\n",
      "kwargs: {}\n",
      "----------------------------debug base_layer.py 1009\n",
      "----------------------------debug base_layer.py 1011\n",
      "----------------------------debug base_layer.py 1013\n",
      "----------------------------debug base_layer.py 1023\n",
      "The caller function is '_functional_construction_call' in module 'keras.engine.base_layer' (current in base_layer.py 2358)\n",
      "The caller function is '<listcomp>' in module 'tensorflow.python.util.nest' (currently in keras_tensor.py 670)\n",
      "----------------------------debug keras_tensor.py 243\n",
      "The caller function is 'keras_tensor_to_placeholder' in module 'keras.engine.keras_tensor' (currently in keras_tensor.py 248)\n",
      "keras_tensor.py self: KerasTensor(type_spec=TensorSpec(shape=(None, 128), dtype=tf.float32, name=None), name='dense/Relu:0', description=\"created by layer 'dense'\")\n",
      "component_to_placeholder: <function KerasTensor._to_placeholder.<locals>.component_to_placeholder at 0x0000013B7E2BBAC0>\n",
      "self.type_spec: TensorSpec(shape=(None, 128), dtype=tf.float32, name=None)\n",
      "The caller function is '<listcomp>' in module 'tensorflow.python.util.nest' (current in keras.engine.keras_tensor 238)\n",
      "The caller function is 'placeholder' in module 'tensorflow.python.ops.array_ops' (current in gen_array_ops.py 6902)\n",
      "name: None\n",
      "dtype: 1\n",
      "The caller function is '<listcomp>' in module 'tensorflow.python.util.nest' (currently in keras_tensor.py 670)\n",
      "----------------------------debug base_layer.py 2441\n",
      "----------------------------debug traceback_utils.py 101\n",
      "The caller function is '_infer_output_signature' in module 'keras.engine.base_layer' (current in base_layer.py 989)\n",
      "----------------------------debug base_layer.py 2462\n",
      "----------------------------debug base_layer.py 2471\n",
      "----------------------------debug base_layer.py 983\n",
      "str(self.name): dense_1\n",
      "The caller function is 'error_handler' in module 'keras.utils.traceback_utils' (current in base_layer.py 989)\n",
      "args: (<KerasTensor: shape=(None, 128) dtype=float32 (created by layer 'leaky_re_lu')>,)\n",
      "kwargs: {}\n",
      "----------------------------debug base_layer.py 1009\n",
      "----------------------------debug base_layer.py 1011\n",
      "----------------------------debug base_layer.py 1013\n",
      "----------------------------debug base_layer.py 1023\n",
      "The caller function is '_functional_construction_call' in module 'keras.engine.base_layer' (current in base_layer.py 2358)\n",
      "The caller function is '<listcomp>' in module 'tensorflow.python.util.nest' (currently in keras_tensor.py 670)\n",
      "----------------------------debug keras_tensor.py 243\n",
      "The caller function is 'keras_tensor_to_placeholder' in module 'keras.engine.keras_tensor' (currently in keras_tensor.py 248)\n",
      "keras_tensor.py self: KerasTensor(type_spec=TensorSpec(shape=(None, 128), dtype=tf.float32, name=None), name='leaky_re_lu/LeakyRelu:0', description=\"created by layer 'leaky_re_lu'\")\n",
      "component_to_placeholder: <function KerasTensor._to_placeholder.<locals>.component_to_placeholder at 0x0000013B7E2BBB50>\n",
      "self.type_spec: TensorSpec(shape=(None, 128), dtype=tf.float32, name=None)\n",
      "The caller function is '<listcomp>' in module 'tensorflow.python.util.nest' (current in keras.engine.keras_tensor 238)\n",
      "The caller function is 'placeholder' in module 'tensorflow.python.ops.array_ops' (current in gen_array_ops.py 6902)\n",
      "name: None\n",
      "dtype: 1\n",
      "The caller function is '<listcomp>' in module 'tensorflow.python.util.nest' (currently in keras_tensor.py 670)\n",
      "----------------------------debug base_layer.py 2441\n",
      "----------------------------debug traceback_utils.py 101\n",
      "The caller function is '_infer_output_signature' in module 'keras.engine.base_layer' (current in base_layer.py 989)\n",
      "----------------------------debug base_layer.py 2462\n",
      "----------------------------debug base_layer.py 2471\n",
      "----------------------------debug base_layer.py 983\n",
      "str(self.name): leaky_re_lu_1\n",
      "The caller function is 'error_handler' in module 'keras.utils.traceback_utils' (current in base_layer.py 989)\n",
      "args: (<KerasTensor: shape=(None, 64) dtype=float32 (created by layer 'dense_1')>,)\n",
      "kwargs: {}\n",
      "----------------------------debug base_layer.py 1009\n",
      "----------------------------debug base_layer.py 1011\n",
      "----------------------------debug base_layer.py 1013\n",
      "----------------------------debug base_layer.py 1023\n",
      "The caller function is '_functional_construction_call' in module 'keras.engine.base_layer' (current in base_layer.py 2358)\n",
      "The caller function is '<listcomp>' in module 'tensorflow.python.util.nest' (currently in keras_tensor.py 670)\n",
      "----------------------------debug keras_tensor.py 243\n",
      "The caller function is 'keras_tensor_to_placeholder' in module 'keras.engine.keras_tensor' (currently in keras_tensor.py 248)\n",
      "keras_tensor.py self: KerasTensor(type_spec=TensorSpec(shape=(None, 64), dtype=tf.float32, name=None), name='dense_1/Relu:0', description=\"created by layer 'dense_1'\")\n",
      "component_to_placeholder: <function KerasTensor._to_placeholder.<locals>.component_to_placeholder at 0x0000013B7E2BBD00>\n",
      "self.type_spec: TensorSpec(shape=(None, 64), dtype=tf.float32, name=None)\n",
      "The caller function is '<listcomp>' in module 'tensorflow.python.util.nest' (current in keras.engine.keras_tensor 238)\n",
      "The caller function is 'placeholder' in module 'tensorflow.python.ops.array_ops' (current in gen_array_ops.py 6902)\n",
      "name: None\n",
      "dtype: 1\n",
      "The caller function is '<listcomp>' in module 'tensorflow.python.util.nest' (currently in keras_tensor.py 670)\n",
      "----------------------------debug base_layer.py 2441\n",
      "----------------------------debug traceback_utils.py 101\n",
      "The caller function is '_infer_output_signature' in module 'keras.engine.base_layer' (current in base_layer.py 989)\n",
      "----------------------------debug base_layer.py 2462\n",
      "----------------------------debug base_layer.py 2471\n",
      "----------------------------debug base_layer.py 983\n",
      "str(self.name): dense_2\n",
      "The caller function is 'error_handler' in module 'keras.utils.traceback_utils' (current in base_layer.py 989)\n",
      "args: (<KerasTensor: shape=(None, 64) dtype=float32 (created by layer 'leaky_re_lu_1')>,)\n",
      "kwargs: {}\n",
      "----------------------------debug base_layer.py 1009\n",
      "----------------------------debug base_layer.py 1011\n",
      "----------------------------debug base_layer.py 1013\n",
      "----------------------------debug base_layer.py 1023\n",
      "The caller function is '_functional_construction_call' in module 'keras.engine.base_layer' (current in base_layer.py 2358)\n",
      "The caller function is '<listcomp>' in module 'tensorflow.python.util.nest' (currently in keras_tensor.py 670)\n",
      "----------------------------debug keras_tensor.py 243\n",
      "The caller function is 'keras_tensor_to_placeholder' in module 'keras.engine.keras_tensor' (currently in keras_tensor.py 248)\n",
      "keras_tensor.py self: KerasTensor(type_spec=TensorSpec(shape=(None, 64), dtype=tf.float32, name=None), name='leaky_re_lu_1/LeakyRelu:0', description=\"created by layer 'leaky_re_lu_1'\")\n",
      "component_to_placeholder: <function KerasTensor._to_placeholder.<locals>.component_to_placeholder at 0x0000013B7E2BBAC0>\n",
      "self.type_spec: TensorSpec(shape=(None, 64), dtype=tf.float32, name=None)\n",
      "The caller function is '<listcomp>' in module 'tensorflow.python.util.nest' (current in keras.engine.keras_tensor 238)\n",
      "The caller function is 'placeholder' in module 'tensorflow.python.ops.array_ops' (current in gen_array_ops.py 6902)\n",
      "name: None\n",
      "dtype: 1\n",
      "The caller function is '<listcomp>' in module 'tensorflow.python.util.nest' (currently in keras_tensor.py 670)\n",
      "----------------------------debug base_layer.py 2441\n",
      "----------------------------debug traceback_utils.py 101\n",
      "The caller function is '_infer_output_signature' in module 'keras.engine.base_layer' (current in base_layer.py 989)\n",
      "----------------------------debug base_layer.py 2462\n",
      "----------------------------debug base_layer.py 2471\n",
      "----------------------------debug base_layer.py 983\n",
      "str(self.name): leaky_re_lu_2\n",
      "The caller function is 'error_handler' in module 'keras.utils.traceback_utils' (current in base_layer.py 989)\n",
      "args: (<KerasTensor: shape=(None, 32) dtype=float32 (created by layer 'dense_2')>,)\n",
      "kwargs: {}\n",
      "----------------------------debug base_layer.py 1009\n",
      "----------------------------debug base_layer.py 1011\n",
      "----------------------------debug base_layer.py 1013\n",
      "----------------------------debug base_layer.py 1023\n",
      "The caller function is '_functional_construction_call' in module 'keras.engine.base_layer' (current in base_layer.py 2358)\n",
      "The caller function is '<listcomp>' in module 'tensorflow.python.util.nest' (currently in keras_tensor.py 670)\n",
      "----------------------------debug keras_tensor.py 243\n",
      "The caller function is 'keras_tensor_to_placeholder' in module 'keras.engine.keras_tensor' (currently in keras_tensor.py 248)\n",
      "keras_tensor.py self: KerasTensor(type_spec=TensorSpec(shape=(None, 32), dtype=tf.float32, name=None), name='dense_2/Relu:0', description=\"created by layer 'dense_2'\")\n",
      "component_to_placeholder: <function KerasTensor._to_placeholder.<locals>.component_to_placeholder at 0x0000013B7E2BBB50>\n",
      "self.type_spec: TensorSpec(shape=(None, 32), dtype=tf.float32, name=None)\n",
      "The caller function is '<listcomp>' in module 'tensorflow.python.util.nest' (current in keras.engine.keras_tensor 238)\n",
      "The caller function is 'placeholder' in module 'tensorflow.python.ops.array_ops' (current in gen_array_ops.py 6902)\n",
      "name: None\n",
      "dtype: 1\n",
      "The caller function is '<listcomp>' in module 'tensorflow.python.util.nest' (currently in keras_tensor.py 670)\n",
      "----------------------------debug base_layer.py 2441\n",
      "----------------------------debug traceback_utils.py 101\n",
      "The caller function is '_infer_output_signature' in module 'keras.engine.base_layer' (current in base_layer.py 989)\n",
      "----------------------------debug base_layer.py 2462\n",
      "----------------------------debug base_layer.py 2471\n",
      "----------------------------debug base_layer.py 983\n",
      "str(self.name): dense_3\n",
      "The caller function is 'error_handler' in module 'keras.utils.traceback_utils' (current in base_layer.py 989)\n",
      "args: (<KerasTensor: shape=(None, 32) dtype=float32 (created by layer 'leaky_re_lu_2')>,)\n",
      "kwargs: {}\n",
      "----------------------------debug base_layer.py 1009\n",
      "----------------------------debug base_layer.py 1011\n",
      "----------------------------debug base_layer.py 1013\n",
      "----------------------------debug base_layer.py 1023\n",
      "The caller function is '_functional_construction_call' in module 'keras.engine.base_layer' (current in base_layer.py 2358)\n",
      "The caller function is '<listcomp>' in module 'tensorflow.python.util.nest' (currently in keras_tensor.py 670)\n",
      "----------------------------debug keras_tensor.py 243\n",
      "The caller function is 'keras_tensor_to_placeholder' in module 'keras.engine.keras_tensor' (currently in keras_tensor.py 248)\n",
      "keras_tensor.py self: KerasTensor(type_spec=TensorSpec(shape=(None, 32), dtype=tf.float32, name=None), name='leaky_re_lu_2/LeakyRelu:0', description=\"created by layer 'leaky_re_lu_2'\")\n",
      "component_to_placeholder: <function KerasTensor._to_placeholder.<locals>.component_to_placeholder at 0x0000013B7E2BBD00>\n",
      "self.type_spec: TensorSpec(shape=(None, 32), dtype=tf.float32, name=None)\n",
      "The caller function is '<listcomp>' in module 'tensorflow.python.util.nest' (current in keras.engine.keras_tensor 238)\n",
      "The caller function is 'placeholder' in module 'tensorflow.python.ops.array_ops' (current in gen_array_ops.py 6902)\n",
      "name: None\n",
      "dtype: 1\n",
      "The caller function is '<listcomp>' in module 'tensorflow.python.util.nest' (currently in keras_tensor.py 670)\n",
      "----------------------------debug base_layer.py 2441\n",
      "----------------------------debug traceback_utils.py 101\n",
      "The caller function is '_infer_output_signature' in module 'keras.engine.base_layer' (current in base_layer.py 989)\n",
      "----------------------------debug base_layer.py 2462\n",
      "----------------------------debug base_layer.py 2471\n",
      "----------------------------debug base_layer.py 983\n",
      "str(self.name): leaky_re_lu_3\n",
      "The caller function is 'error_handler' in module 'keras.utils.traceback_utils' (current in base_layer.py 989)\n",
      "args: (<KerasTensor: shape=(None, 4) dtype=float32 (created by layer 'dense_3')>,)\n",
      "kwargs: {}\n",
      "----------------------------debug base_layer.py 1009\n",
      "----------------------------debug base_layer.py 1011\n",
      "----------------------------debug base_layer.py 1013\n",
      "----------------------------debug base_layer.py 1023\n",
      "The caller function is '_functional_construction_call' in module 'keras.engine.base_layer' (current in base_layer.py 2358)\n",
      "The caller function is '<listcomp>' in module 'tensorflow.python.util.nest' (currently in keras_tensor.py 670)\n",
      "----------------------------debug keras_tensor.py 243\n",
      "The caller function is 'keras_tensor_to_placeholder' in module 'keras.engine.keras_tensor' (currently in keras_tensor.py 248)\n",
      "keras_tensor.py self: KerasTensor(type_spec=TensorSpec(shape=(None, 4), dtype=tf.float32, name=None), name='dense_3/Relu:0', description=\"created by layer 'dense_3'\")\n",
      "component_to_placeholder: <function KerasTensor._to_placeholder.<locals>.component_to_placeholder at 0x0000013B7E2BBAC0>\n",
      "self.type_spec: TensorSpec(shape=(None, 4), dtype=tf.float32, name=None)\n",
      "The caller function is '<listcomp>' in module 'tensorflow.python.util.nest' (current in keras.engine.keras_tensor 238)\n",
      "The caller function is 'placeholder' in module 'tensorflow.python.ops.array_ops' (current in gen_array_ops.py 6902)\n",
      "name: None\n",
      "dtype: 1\n",
      "The caller function is '<listcomp>' in module 'tensorflow.python.util.nest' (currently in keras_tensor.py 670)\n",
      "----------------------------debug base_layer.py 2441\n",
      "----------------------------debug traceback_utils.py 101\n",
      "The caller function is '_infer_output_signature' in module 'keras.engine.base_layer' (current in base_layer.py 989)\n",
      "----------------------------debug base_layer.py 2462\n",
      "----------------------------debug base_layer.py 2471\n",
      "----------------------------debug functional.py 1102\n",
      "inputs: [<KerasTensor: shape=(None, 224, 224, 3) dtype=float32 (created by layer 'input_2')>]\n",
      "The caller function is '_init_graph_network' in module 'keras.engine.functional' (current in functional.py 1108)\n",
      "----------------------------debug functional.py 1159\n",
      "len nodes_by_depth: 11\n",
      "layers_with_complete_input: ['vgg19', 'flatten', 'dense', 'leaky_re_lu', 'dense_1', 'leaky_re_lu_1', 'dense_2', 'leaky_re_lu_2', 'dense_3', 'leaky_re_lu_3']\n"
     ]
    }
   ],
   "source": [
    "from tensorflow.keras.applications import VGG19\n",
    "from tensorflow.keras.layers import Input, Conv2D\n",
    "from tensorflow.keras.models import Model\n",
    "from tensorflow.keras.optimizers import Adam\n",
    "from tensorflow.keras.preprocessing.image import ImageDataGenerator\n",
    "from tensorflow.keras.callbacks import ReduceLROnPlateau, EarlyStopping\n",
    "from tensorflow.keras.losses import Huber\n",
    "from keras.layers import Flatten, Dense, LeakyReLU\n",
    "from tensorflow.keras.losses import mse\n",
    "\n",
    "# Load the pre-trained ResNet50 model\n",
    "base_model = VGG19(weights='imagenet', include_top=False, input_shape=(224, 224, 3))\n",
    "\n",
    "# Add bbox output layer\n",
    "inputs = Input(shape=(224, 224, 3))\n",
    "x = base_model(inputs)\n",
    "\n",
    "# flatten the max-pooling output of ResNet50\n",
    "flatten = Flatten()(x)\n",
    "# construct a fully-connected layer header to output the predicted\n",
    "# bounding box coordinates\n",
    "bboxHead = Dense(128, activation=\"relu\")(flatten)\n",
    "bboxHead = LeakyReLU(alpha=0.1)(bboxHead)\n",
    "bboxHead = Dense(64, activation=\"relu\")(bboxHead)\n",
    "bboxHead = LeakyReLU(alpha=0.1)(bboxHead)\n",
    "bboxHead = Dense(32, activation=\"relu\")(bboxHead)\n",
    "bboxHead = LeakyReLU(alpha=0.1)(bboxHead)\n",
    "bboxHead = Dense(4, activation=\"relu\")(bboxHead)\n",
    "bboxHead = LeakyReLU(alpha=0.1)(bboxHead)\n",
    "\n",
    "# construct the model we will fine-tune for bounding box regression\n",
    "model = Model(inputs=inputs, outputs=bboxHead)\n",
    "\n",
    "# Freeze the weights of the pre-trained layers\n",
    "for layer in base_model.layers:\n",
    "    layer.trainable = True\n",
    "\n",
    "\n",
    "# Compile the model\n",
    "optimizer = Adam(learning_rate=1e-4)\n",
    "model.compile(optimizer=optimizer, loss=mse)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "c538213e",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "----------------------------debug functional.py 1102\n",
      "inputs: [<KerasTensor: shape=(None, 784) dtype=float32 (created by layer 'input_2')>]\n",
      "The caller function is '_init_graph_network' in module 'keras.engine.functional' (current in functional.py 1108)\n",
      "----------------------------debug functional.py 1159\n",
      "len nodes_by_depth: 1\n",
      "layers_with_complete_input: []\n",
      "----------------------------debug base_layer.py 983\n",
      "str(self.name): dense_3\n",
      "The caller function is 'error_handler' in module 'keras.utils.traceback_utils' (current in base_layer.py 989)\n",
      "args: (<KerasTensor: shape=(None, 784) dtype=float32 (created by layer 'input_2')>,)\n",
      "kwargs: {}\n",
      "----------------------------debug base_layer.py 1009\n",
      "----------------------------debug base_layer.py 1011\n",
      "----------------------------debug base_layer.py 1013\n",
      "----------------------------debug base_layer.py 1023\n",
      "The caller function is '_functional_construction_call' in module 'keras.engine.base_layer' (current in base_layer.py 2358)\n",
      "The caller function is '<listcomp>' in module 'tensorflow.python.util.nest' (currently in keras_tensor.py 670)\n",
      "----------------------------debug keras_tensor.py 243\n",
      "The caller function is 'keras_tensor_to_placeholder' in module 'keras.engine.keras_tensor' (currently in keras_tensor.py 248)\n",
      "keras_tensor.py self: KerasTensor(type_spec=TensorSpec(shape=(None, 784), dtype=tf.float32, name='input_2'), name='input_2', description=\"created by layer 'input_2'\")\n",
      "component_to_placeholder: <function KerasTensor._to_placeholder.<locals>.component_to_placeholder at 0x000001B379429510>\n",
      "self.type_spec: TensorSpec(shape=(None, 784), dtype=tf.float32, name='input_2')\n",
      "The caller function is '<listcomp>' in module 'tensorflow.python.util.nest' (current in keras.engine.keras_tensor 238)\n",
      "The caller function is 'placeholder' in module 'tensorflow.python.ops.array_ops' (current in gen_array_ops.py 6902)\n",
      "name: None\n",
      "dtype: 1\n",
      "The caller function is '<listcomp>' in module 'tensorflow.python.util.nest' (currently in keras_tensor.py 670)\n",
      "----------------------------debug base_layer.py 2441\n",
      "----------------------------debug traceback_utils.py 101\n",
      "The caller function is '_infer_output_signature' in module 'keras.engine.base_layer' (current in base_layer.py 989)\n",
      "----------------------------debug base_layer.py 2462\n",
      "----------------------------debug base_layer.py 2471\n",
      "----------------------------debug functional.py 1102\n",
      "inputs: [<KerasTensor: shape=(None, 784) dtype=float32 (created by layer 'input_2')>]\n",
      "The caller function is '_init_graph_network' in module 'keras.engine.functional' (current in functional.py 1108)\n",
      "----------------------------debug functional.py 1159\n",
      "len nodes_by_depth: 2\n",
      "layers_with_complete_input: ['dense_3']\n",
      "----------------------------debug base_layer.py 983\n",
      "str(self.name): dense_4\n",
      "The caller function is 'error_handler' in module 'keras.utils.traceback_utils' (current in base_layer.py 989)\n",
      "args: (<KerasTensor: shape=(None, 128) dtype=float32 (created by layer 'dense_3')>,)\n",
      "kwargs: {}\n",
      "----------------------------debug base_layer.py 1009\n",
      "----------------------------debug base_layer.py 1011\n",
      "----------------------------debug base_layer.py 1013\n",
      "----------------------------debug base_layer.py 1023\n",
      "The caller function is '_functional_construction_call' in module 'keras.engine.base_layer' (current in base_layer.py 2358)\n",
      "The caller function is '<listcomp>' in module 'tensorflow.python.util.nest' (currently in keras_tensor.py 670)\n",
      "----------------------------debug keras_tensor.py 243\n",
      "The caller function is 'keras_tensor_to_placeholder' in module 'keras.engine.keras_tensor' (currently in keras_tensor.py 248)\n",
      "keras_tensor.py self: KerasTensor(type_spec=TensorSpec(shape=(None, 128), dtype=tf.float32, name=None), name='dense_3/Relu:0', description=\"created by layer 'dense_3'\")\n",
      "component_to_placeholder: <function KerasTensor._to_placeholder.<locals>.component_to_placeholder at 0x000001B379169D80>\n",
      "self.type_spec: TensorSpec(shape=(None, 128), dtype=tf.float32, name=None)\n",
      "The caller function is '<listcomp>' in module 'tensorflow.python.util.nest' (current in keras.engine.keras_tensor 238)\n",
      "The caller function is 'placeholder' in module 'tensorflow.python.ops.array_ops' (current in gen_array_ops.py 6902)\n",
      "name: None\n",
      "dtype: 1\n",
      "The caller function is '<listcomp>' in module 'tensorflow.python.util.nest' (currently in keras_tensor.py 670)\n",
      "----------------------------debug base_layer.py 2441\n",
      "----------------------------debug traceback_utils.py 101\n",
      "The caller function is '_infer_output_signature' in module 'keras.engine.base_layer' (current in base_layer.py 989)\n",
      "----------------------------debug base_layer.py 2462\n",
      "----------------------------debug base_layer.py 2471\n",
      "----------------------------debug functional.py 1102\n",
      "inputs: [<KerasTensor: shape=(None, 784) dtype=float32 (created by layer 'input_2')>]\n",
      "The caller function is '_init_graph_network' in module 'keras.engine.functional' (current in functional.py 1108)\n",
      "----------------------------debug functional.py 1159\n",
      "len nodes_by_depth: 3\n",
      "layers_with_complete_input: ['dense_3', 'dense_4']\n",
      "----------------------------debug base_layer.py 983\n",
      "str(self.name): dense_5\n",
      "The caller function is 'error_handler' in module 'keras.utils.traceback_utils' (current in base_layer.py 989)\n",
      "args: (<KerasTensor: shape=(None, 64) dtype=float32 (created by layer 'dense_4')>,)\n",
      "kwargs: {}\n",
      "----------------------------debug base_layer.py 1009\n",
      "----------------------------debug base_layer.py 1011\n",
      "----------------------------debug base_layer.py 1013\n",
      "----------------------------debug base_layer.py 1023\n",
      "The caller function is '_functional_construction_call' in module 'keras.engine.base_layer' (current in base_layer.py 2358)\n",
      "The caller function is '<listcomp>' in module 'tensorflow.python.util.nest' (currently in keras_tensor.py 670)\n",
      "----------------------------debug keras_tensor.py 243\n",
      "The caller function is 'keras_tensor_to_placeholder' in module 'keras.engine.keras_tensor' (currently in keras_tensor.py 248)\n",
      "keras_tensor.py self: KerasTensor(type_spec=TensorSpec(shape=(None, 64), dtype=tf.float32, name=None), name='dense_4/Relu:0', description=\"created by layer 'dense_4'\")\n",
      "component_to_placeholder: <function KerasTensor._to_placeholder.<locals>.component_to_placeholder at 0x000001B37916A8C0>\n",
      "self.type_spec: TensorSpec(shape=(None, 64), dtype=tf.float32, name=None)\n",
      "The caller function is '<listcomp>' in module 'tensorflow.python.util.nest' (current in keras.engine.keras_tensor 238)\n",
      "The caller function is 'placeholder' in module 'tensorflow.python.ops.array_ops' (current in gen_array_ops.py 6902)\n",
      "name: None\n",
      "dtype: 1\n",
      "The caller function is '<listcomp>' in module 'tensorflow.python.util.nest' (currently in keras_tensor.py 670)\n",
      "----------------------------debug base_layer.py 2441\n",
      "----------------------------debug traceback_utils.py 101\n",
      "The caller function is '_infer_output_signature' in module 'keras.engine.base_layer' (current in base_layer.py 989)\n",
      "----------------------------debug base_layer.py 2462\n",
      "----------------------------debug base_layer.py 2471\n",
      "----------------------------debug functional.py 1102\n",
      "inputs: [<KerasTensor: shape=(None, 784) dtype=float32 (created by layer 'input_2')>]\n",
      "The caller function is '_init_graph_network' in module 'keras.engine.functional' (current in functional.py 1108)\n",
      "----------------------------debug functional.py 1159\n",
      "len nodes_by_depth: 4\n",
      "layers_with_complete_input: ['dense_3', 'dense_4', 'dense_5']\n",
      "----------------------------debug functional.py 1102\n",
      "inputs: [<KerasTensor: shape=(None, 784) dtype=float32 (created by layer 'input_2')>]\n",
      "The caller function is '_init_graph_network' in module 'keras.engine.functional' (current in functional.py 1108)\n",
      "----------------------------debug functional.py 1159\n",
      "len nodes_by_depth: 4\n",
      "layers_with_complete_input: ['dense_3', 'dense_4', 'dense_5']\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "<KerasTensor: shape=(None, 784) dtype=float32 (created by layer 'input_2')>"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import tensorflow as tf\n",
    "from tensorflow import keras\n",
    "\n",
    "# Create a sample model\n",
    "model = keras.Sequential([\n",
    "    keras.layers.Input(shape=(784,)),\n",
    "    keras.layers.Dense(128, activation='relu'),\n",
    "    keras.layers.Dense(64, activation='relu'),\n",
    "    keras.layers.Dense(10, activation='softmax')\n",
    "])\n",
    "\n",
    "# Assume 'output_layer' is the last layer in the model\n",
    "output_layer = model.layers[-1]\n",
    "\n",
    "# Create a functional model to access intermediate layers\n",
    "model_with_intermediates = keras.Model(inputs=model.input, outputs=output_layer.output)\n",
    "\n",
    "# Start tracing back from the output layer\n",
    "layers_backtracked = [output_layer]\n",
    "model_with_intermediates.input\n",
    "\n",
    "while layers_backtracked[-1] != model_with_intermediates.input:\n",
    "    prev_layer_output = layers_backtracked[-1].input\n",
    "    prev_layer = prev_layer_output.op.inputs[0].owner.layer\n",
    "    layers_backtracked.append(prev_layer)\n",
    "\n",
    "# # Reverse the list to get the tracing from input to output\n",
    "# layers_backtracked.reverse()\n",
    "\n",
    "# # Print the layers traced back\n",
    "# for layer in layers_backtracked:\n",
    "#     print(layer)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "5593c89a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'dense_5'"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "output_layer.name"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "a9d3f2fd",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<KerasTensor: shape=(None, 64) dtype=float32 (created by layer 'dense_4')>"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "layers_backtracked[-1].input"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "f84baa5c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<KerasTensor: shape=(None, 784) dtype=float32 (created by layer 'input_2')>"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "model.layers[0].input"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "8988e085",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<KerasTensor: shape=(None, 128) dtype=float32 (created by layer 'dense_3')>"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "model.layers[1].input"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "a7955cfe",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Moviepy - Building video predection_sample.mp4.\n",
      "Moviepy - Writing video predection_sample.mp4\n",
      "\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "                                                                                                                       \r"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Moviepy - Done !\n",
      "Moviepy - video ready predection_sample.mp4\n"
     ]
    }
   ],
   "source": [
    "from moviepy.editor import VideoFileClip\n",
    "\n",
    "# Specify the paths for the input AVI file and output MP4 file\n",
    "input_avi_path = 'df_dist_augmentation.avi'\n",
    "output_mp4_path = 'predection_sample.mp4'\n",
    "\n",
    "# Load the AVI video clip\n",
    "video_clip = VideoFileClip(input_avi_path)\n",
    "\n",
    "# Write the video clip to an MP4 file\n",
    "video_clip.write_videofile(output_mp4_path, codec='libx264')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "b8e52d6e",
   "metadata": {},
   "outputs": [],
   "source": [
    "# COCO Class names\n",
    "# Index of the class in the list is its ID. For example, to get ID of\n",
    "# the teddy bear class, use: class_names.index('teddy bear')\n",
    "class_names = ['BG', 'person', 'bicycle', 'car', 'motorcycle', 'airplane',\n",
    "               'bus', 'train', 'truck', 'boat', 'traffic light',\n",
    "               'fire hydrant', 'stop sign', 'parking meter', 'bench', 'bird',\n",
    "               'cat', 'dog', 'horse', 'sheep', 'cow', 'elephant', 'bear',\n",
    "               'zebra', 'giraffe', 'backpack', 'umbrella', 'handbag', 'tie',\n",
    "               'suitcase', 'frisbee', 'skis', 'snowboard', 'sports ball',\n",
    "               'kite', 'baseball bat', 'baseball glove', 'skateboard',\n",
    "               'surfboard', 'tennis racket', 'bottle', 'wine glass', 'cup',\n",
    "               'fork', 'knife', 'spoon', 'bowl', 'banana', 'apple',\n",
    "               'sandwich', 'orange', 'broccoli', 'carrot', 'hot dog', 'pizza',\n",
    "               'donut', 'cake', 'chair', 'couch', 'potted plant', 'bed',\n",
    "               'dining table', 'toilet', 'tv', 'laptop', 'mouse', 'remote',\n",
    "               'keyboard', 'cell phone', 'microwave', 'oven', 'toaster',\n",
    "               'sink', 'refrigerator', 'book', 'clock', 'vase', 'scissors',\n",
    "               'teddy bear', 'hair drier', 'toothbrush']\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "64b6ddcf",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'person'"
      ]
     },
     "execution_count": 27,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "class_names[1]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "586f029f",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
